// @ton/tolk-js/dist/stdlib.tolk.js
// stdlib files are stored as strings in order to make it work on web (without fs.readFile)
module.exports = {

'@stdlib/common.tolk':`// Standard library for Tolk (LGPL licence).
// It contains common functions that are available out of the box, the user doesn't have to import anything.
// More specific functions are required to be imported explicitly, like "@stdlib/gas-payments".
tolk 1.3

/**
    Built-in types.
 */

/// \`int\` is the primitive 257-bit signed integer type.
type int = builtin

/// \`bool\` is a classic boolean type, which can hold only two values: \`true\` and \`false\`.
/// At the TVM (TON virtual machine) level, it's an integer -1 or 0.
/// Note: \`boolVar as int\` is possible, but remember, that true is -1, not 1!
type bool = builtin

/// \`cell\` is a data structure, which can hold of up to 1023 bits (not bytes!)
/// and up to 4 references (refs) to other cells.
/// Both contract code and contract state are represented by a tree of cells.
/// [See docs](https://docs.ton.org/v3/documentation/data-formats/tlb/cell-boc)
type cell = builtin

/// \`slice\` is a "cell opened for reading".
/// When you call [cell.beginParse], you get a \`slice\`, from which you can load binary data
/// or high-level structures with [T.fromSlice].
type slice = builtin

/// \`builder\` is a "cell at the stage of creation".
/// When you call [beginCell], you get a \`builder\`, populate it with binary data or structures,
/// and after [builder.endCell], you get a \`cell\`.
type builder = builtin

/// \`continuation\` is an "executable cell" representing executable TVM bytecode.
/// They are used to manage execution flow in TVM programs and serve as
/// the basis for function calls, exception handling, and control flow operations. 
type continuation = builtin

/// \`address\` represents a standard (internal) address: workchain + hash.
/// Addresses are comparable with \`==\` operator, they have methods like [address.getWorkchain].
/// In binary, it's 267 bits: '10' (addr_std) + '0' (no anycast) + workchain (int8) + hash (uint256).
/// If you wish to describe "a potentially missing address", use \`address?\` (nullable).
/// When set to \`null\`, \`address?\` will be serialized as '00' ("none address"):
/// '00' (two zero bits) is a standard representation of a "missing" ("none") address.
/// At the TVM level, addresses are slices, so it's possible to  \`someAddress as slice\`.
/// [See docs](https://docs.ton.org/learn/overviews/addresses#address-of-smart-contract)
type address = builtin

/// \`any_address\` represents an internal/external/none address.
/// You can test it with [any_address.isExternal] and similar methods.
/// If you wish to describe "internal or none address", use \`address?\` (nullable)
/// instead of "any". That's why \`any_address\` is used in practice extremely rarely.
/// At the TVM level, addresses are slices, so it's possible to  \`anyAddress as slice\`.
/// [See docs](https://docs.ton.org/learn/overviews/addresses#address-of-smart-contract)
type any_address = builtin

/// \`never\` is a special type that represents computations that never complete normally.
/// A function that always throws an exception returns \`never\`, and the compiler knows
/// that any code after it is unreachable.
type never = builtin

/// \`int8\`, \`int32\`, \`int222\`, etc. is "a fixed-width signed integer with N bits", N <= 257.
/// Note: it's still \`int\` at runtime, you can assign "100500" to "int8":
/// overflow will happen at serialization to a cell/builder, NOT at assignment. 
type intN = builtin

/// \`uint32\`, \`uint64\`, \`uint111\`, etc. is "a fixed-width unsigned integer with N bits", N <= 256.
/// Note: it's still \`int\` at runtime, you can assign "100500" to "uint8":
/// overflow will happen at serialization to a cell/builder, NOT at assignment. 
type uintN = builtin

/// \`coins\` is a special primitive representing "nanotoncoins". One TON = 10^9 nanotoncoins.
/// You can create coins with \`ton()\` function: \`ton("0.05")\` (actually, \`int\` 50000000 at runtime).
/// Arithmetic operations on \`coins\` degrade to 257-bit \`int\` type.
type coins = builtin

/// \`varint16\` is \`int\` at runtime, but serialized as "variadic signed int", -2^119 <= X < 2^119. 
type varint16 = builtin

/// \`varuint16\` is \`int\` at runtime, but serialized as "variadic unsigned int", 0 <= X < 2^120. 
type varuint16 = builtin

/// \`varint32\` is \`int\` at runtime, but serialized as "variadic signed int", -2^247 <= X < 2^247. 
type varint32 = builtin

/// \`varuint32\` is \`int\` at runtime, but serialized as "variadic unsigned int", 0 <= X < 2^248. 
type varuint32 = builtin

/// \`string\` represents text data stored as a cell in snake format.
/// Snake format: data bytes, if more data exists — ref to next cell.
/// Example of "abcd" in a single chunk: cell(data: "abcd") (32 bits inside)
/// Example of "abcd" in two chunks: cell(data: "ab", ref: cell(data: "cd"))
/// At the TVM level, \`string\` is a TVM \`CELL\` (1 stack slot).
type string = builtin

/// \`bits256\`, \`bits111\`, etc. is "a fixed-width slice with N bits and 0 refs", N <= 1023.
/// Note: use \`as\` operator to convert \`slice\` to \`bitsN\`: \`someSlice as bits256\` 
/// (manually writing \`as\` enforces you to think that this conversion is correct).
/// Note: similar to \`intN\`, you can assign an invalid slice to \`bitsN\`,
/// an error will be fired at serialization with [T.toCell] and similar, NOT at assignment.
type bitsN = builtin

/// \`bytes8\`, \`bytes99\`, etc. is a convenient alias for \`bits(N*8)\`
type bytesN = builtin

/// \`map<K, V>\` is "a map from a key K to a value V".
/// Internally, it's an "optional cell": an empty map is \`null\`, a non-empty points to a root cell.
/// Restrictions for K and V types:
/// - a key must be fixed-width; valid: \`int32\`, \`uint64\`, \`address\`, \`bits256\`, \`Point\`; invalid: \`int\`, \`coins\`
/// - a value must be serializable; valid: \`int32\`, \`coins\`, \`AnyStruct\`, \`Cell<AnyStruct>\`; invalid: \`int\`, \`builder\`
struct map<K, V> {
    private tvmDict: dict
}

/// \`array<T>\` is a container that stores from 0 to 255 elements of type T.
/// Internally, it's a TVM tuple: the i-th element of a TVM tuple is the i-th element of an array.
/// On a stack, an array occupies one stack slot regardless of its size.
/// \`T\` can be any type (complex types are converted to sub-tuples under the hood).
/// Work with array using its methods: \`arr.push(v)\`, \`arr.get(i)\`, etc.
struct array<T> {
    private tvmTuple: tuple
}

/// \`unknown\` represents one TVM primitive (one stack slot) with a kind unknown at compile-time.
/// Any type \`T\` can be cast to \`unknown\` and back. If T is a TVM primitive, it's a type-only cast.
/// Otherwise, an object is converted to a tuple, and is still stored as one stack slot.
/// \`\`\`
/// var u1 = 5 as unknown;         // stack: "5"
/// u1 as int;                     // stack: "5"
/// var u2 = (10, 20) as unknown;  // stack: "[ 10 20 ]" (a TVM tuple)
/// u2 as (int, int);              // stack: "10 20" (two integers)
/// \`\`\`
/// Be extremely careful using \`unknown\`, because the compiler is unable to validate the casts.
/// An invalid cast from your side leads to stack corruption, resulting in undefined behavior.
type unknown = builtin

/// \`tuple\` represents a TVM tuple. Essentially, it's "an array of unknown elements".
/// Reading an individual element results in \`unknown\`, cast it to perform some actions:
/// \`\`\`
/// // wrong:
/// var elem = someTuple.get(0);    // unknown
/// return elem + 123;              // error, can not apply \`+\`
/// 
/// // correct:
/// var elem = someTuple.get(0) as int;
/// return elem + 123;
/// \`\`\`
type tuple = array<unknown>

/// \`lisp_list<T>\` is a set of nested 2-element TVM tuples:
/// \`[1, [2, [3, null]]]\` represents a list \`[1, 2, 3]\`.
/// Unlike \`array<T>\`, it can represent a list longer than 255 elements,
/// but it has access only to the frontmost element (head).
/// 
/// To use lisp-style list, import the \`@stdlib/lisp-lists\` and call its methods.
/// \`\`\`
/// var list: lisp_list<int> = [];
/// list.prependHead(1);
/// list.prependHead(2);
/// // list is now "2 1" (on a stack: [ 2 [ 1 null ] ])
/// list.getHead(); // 2
/// \`\`\`
struct lisp_list<T> {
    private tvmTuple: [T, lisp_list<T>] | null
}

/// Think of it as "a map with unknown keys and unknown values".
/// Prefer using \`map<K, V>\`, not \`dict\`.
type dict = cell?

/// \`void\` is the unit type representing the absence of a meaningful value.
/// It's similar to both \`void\` and \`unit\` in other languages.
/// Note: a function without return type means "auto infer", NOT "void".
type void = builtin

/// \`Cell<T>\` represents a typed cell reference (as opposed to untyped \`cell\`).
/// \`\`\`
/// struct ExtraData { ... }
///
/// struct MyStorage {
///     ...
///     extra:    Cell<ExtraData>    // TL-B \`^ExtraData\`
///     optional: Cell<ExtraData>?   // TL-B \`(Maybe ^ExtraData)\`
///     code:     cell               // TL-B \`^Cell\`
///     data:     cell?              // TL-B \`(Maybe ^Cell)\`
/// }
/// \`\`\`
/// Note, that \`st = MyStorage.fromSlice(s)\` does NOT deep-load any refs; \`st.extra\` is \`Cell<T>\`, not \`T\`;
/// you should manually call \`st.extra.load()\` to get \`T\` (ExtraData in this example).
struct Cell<T> {
    private readonly tvmCell: cell
}


/**
    Array manipulation primitives.
 */

/// Creates a TVM tuple with zero elements.
/// Deprecated since Tolk v1.3. Use \`[]\` to get an empty array instead: 
/// \`\`\`
/// // old way, deprecated:
/// var t = createEmptyTuple();
/// 
/// // new way, preferred:
/// var t = [];   // array<unknown>
/// 
/// // create an empty typed array:
/// var numbers: array<int> = [];
/// \`\`\`
@pure
@deprecated("use \`[]\` instead of \`createEmptyTuple\`")
fun createEmptyTuple(): array<unknown>
    asm "0 TUPLE"

/// Appends a value to array.
/// If its size exceeds 255, throws a type check exception.
@pure
fun array<T>.push(mutate self, value: T): void
    builtin

/// Returns the first element of a non-empty array.
@pure
fun array<T>.first(self): T
    builtin

/// Returns the \`i\`-th element of an array.
@pure
fun array<T>.get(self, index: int): T
    builtin

/// Sets the \`i\`-th element of an array to a specified value
/// (an element with this index must already exist, a new element is not created).
@pure
fun array<T>.set(mutate self, value: T, index: int): void
    builtin

/// Returns the size of an array (elements count in it).
@pure
fun array<T>.size(self): int
    builtin

/// Returns the last element of a non-empty array.
@pure
fun array<T>.last(self): T
    builtin

/// Pops and returns the last element of a non-empty array.
@pure
fun array<T>.pop(mutate self): T
    builtin

/// Packs any object from a stack to a tuple. 
/// An object occupies N slots on a stack — the tuple will be of size N.
/// It works identically as casting an object to \`unknown\`.
/// Example:
/// \`\`\`
/// struct Point { x: int, y: int }
/// 
/// var p: Point = { x: 1, y: 2 };
/// var t = p.toTuple();     // [ 1 2 ]
/// p = Point.fromTuple(t);  // back
/// t.get(0) as int;         // 1
/// \`\`\`
/// It can be used for logging or low-level purposes like running a child VM or custom "equals".
/// See [T.fromTuple] counterpart.
@pure
fun T.toTuple(self): array<unknown>
    builtin

/// Unpacks a tuple back to an object on a stack.
/// It works identically as casting \`unknown\` to an object.
/// See [T.toTuple] for explanation and examples.
@pure
fun T.fromTuple(packedObject: array<unknown>): T
    builtin

/**
    Mathematical primitives.
 */

/// Converts a constant floating-point string to nanotoncoins.
/// Example: \`ton("0.05")\` is equal to 50000000.
/// Note, that \`ton()\` requires a constant string; \`ton(some_var)\` is an error.
@pure
fun ton(floatString: string): coins
    builtin

/// Computes the minimum of two integers.
@pure
fun min(x: int, y: int): int
    asm "MIN"

/// Computes the maximum of two integers.
@pure
fun max(x: int, y: int): int
    asm "MAX"

/// Sorts two integers.
/// Example: \`minMax(x, y)\` with (x=20, y=10) and with (x=10, y=20) returns (10, 20)
@pure
fun minMax(x: int, y: int): (int, int)
    asm "MINMAX"

/// Computes the absolute value of an integer.
@pure
fun abs(x: int): int
    asm "ABS"

/// Returns the sign of an integer: \`-1\` if x < 0, \`0\` if x == 0, \`1\` if x > 0.
@pure
fun sign(x: int): int
    asm "SGN"

/// Computes the quotient and remainder of x / y. Example: divMod(112,3) = (37,1)
@pure
fun divMod(x: int, y: int): (int, int)
    asm "DIVMOD"

/// Computes the remainder and quotient of x / y. Example: modDiv(112,3) = (1,37)
@pure
fun modDiv(x: int, y: int): (int, int)
    asm(-> 1 0) "DIVMOD"

/// Computes multiple-then-divide: floor(x * y / z).
/// The intermediate result is stored in a 513-bit integer to prevent precision loss.
@pure
fun mulDivFloor(x: int, y: int, z: int): int
    builtin

/// Similar to [mulDivFloor], but rounds the result: round(x * y / z).
@pure
fun mulDivRound(x: int, y: int, z: int): int
    builtin

/// Similar to [mulDivFloor], but ceils the result: ceil(x * y / z).
@pure
fun mulDivCeil(x: int, y: int, z: int): int
    builtin

/// Computes the quotient and remainder of (x * y / z). Example: mulDivMod(112,3,10) = (33,6)
@pure
fun mulDivMod(x: int, y: int, z: int): (int, int)
    builtin


/**
    Global getters and setters of current contract state.
 */

/// \`contract\` is a built-in struct, it has only static methods.
/// Example: \`contract.getCode()\` and other methods.
struct contract

/// Returns the internal address of the current smart contract.
/// If necessary, it can be parsed further using [address.getWorkchain] and others.
@pure
fun contract.getAddress(): address
    asm "MYADDR"

/// Returns the balance (in nanotoncoins) of the smart contract at the start of Computation Phase.
@pure
fun contract.getOriginalBalance(): coins
    asm "BALANCE" "FIRST"

/// Same as [contract.getOriginalBalance], but also returns the balance in extra currencies.
@pure
fun contract.getOriginalBalanceWithExtraCurrencies(): [coins, ExtraCurrenciesMap]
    asm "BALANCE"

/// Returns the persistent contract storage cell. It can be parsed or modified with slice and builder primitives later.
@pure
fun contract.getData(): cell
    asm "c4 PUSH"

/// Sets the persistent contract storage. 
fun contract.setData(c: cell): void
    asm "c4 POP"

/// Retrieves code of smart-contract from c7.
@pure
fun contract.getCode(): cell
    asm "MYCODE"

/// Creates an output action that would change this smart contract code after successful termination of the current run.
fun contract.setCodePostponed(newCode: cell): void
    asm "SETCODE"


/**
    Global getters of current blockchain (environment) state.
*/

const MASTERCHAIN = -1
const BASECHAIN = 0

/// \`blockchain\` is a built-in struct, it has only static methods.
/// Example: \`blockchain.configParam(16)\` and other methods.
struct blockchain

/// Returns current Unix timestamp (in seconds).
@pure
fun blockchain.now(): int
    asm "NOW"

/// Returns the logical time of the current transaction.
@pure
fun blockchain.logicalTime(): int
    asm "LTIME"

/// Returns the starting logical time of the current block.
@pure
fun blockchain.currentBlockLogicalTime(): int
    asm "BLOCKLT"

/// Returns the value of the global configuration parameter with integer index \`i\` as a \`cell\` or \`null\` value.
@pure
fun blockchain.configParam(x: int): cell?
    asm "CONFIGOPTPARAM"

/// Commits current state of registers \`c4\` (persistent data) and \`c5\` (actions)
/// so that the current execution is considered "successful" with the saved values even if an exception
/// in Computation Phase is thrown later.
fun commitContractDataAndActions(): void
    asm "COMMIT"


/**
    Auto packing structures to/from cells.
*/

/// PackOptions allows you to control behavior of \`obj.toCell()\` and similar functions.
struct PackOptions {
    /// when a struct has a field of type \`bits128\` and similar (it's a slice under the hood),
    /// by default, compiler inserts runtime checks (get bits/refs count + compare with 128 + compare with 0);
    /// these checks ensure that serialized binary data will be correct, but they cost gas;
    /// however, if you guarantee that a slice is valid (for example, it comes from trusted sources),
    /// set this option to true to disable runtime checks;
    /// note: \`int32\` and other are always validated for overflow without any extra gas,
    /// so this flag controls only rarely used \`bitsN\` type
    skipBitsNValidation: bool = false
}

/// UnpackOptions allows you to control behavior of \`MyStruct.fromCell(c)\` and similar functions.
struct UnpackOptions {
    /// after finished reading all fields from a cell/slice, call [slice.assertEnd] to ensure no remaining data left;
    /// it's the default behavior, it ensures that you've fully described data you're reading with a struct;
    /// example: \`struct Point { x: int8; y: int8 }\`, input "0102" is ok, "0102FF" will throw excno 9;
    /// note: setting this to false does not decrease gas (DROP from a stack and ENDS cost the same);
    /// note: this option controls [T.fromCell] and [T.fromSlice], but is ignored by [slice.loadAny];
    /// note: \`lazy\` ignores this option, because it reads fields on demand or even skips them
    assertEndAfterReading: bool = true

    /// this excNo is thrown if a prefix doesn't match, e.g. for \`struct (0x01) A\` given input "88...";
    /// similarly, for a union type, this is thrown when none of the opcodes match;
    /// note: \`lazy\` ignores this option if you have \`else\` in \`match\` (you write custom logic there)
    throwIfOpcodeDoesNotMatch: int = 63
}

/// Convert anything to a cell (most likely, you'll call it for structures).
/// \`\`\`
/// var st: MyStorage = { ... };
/// contract.setData(st.toCell());
/// \`\`\`
/// Internally, a builder is created, all fields are serialized one by one, and a builder is flushed
/// (beginCell() + serialize fields + endCell()).
@pure
fun T.toCell(self, options: PackOptions = {}): Cell<T>
    builtin

/// Parse anything from a cell (most likely, you'll call it for structures).
/// \`\`\`
/// var st = MyStorage.fromCell(contract.getData());
/// \`\`\`
/// Internally, a cell is unpacked to a slice, and that slice is parsed
/// (packedCell.beginParse() + read from slice).
@pure
fun T.fromCell(packedCell: cell, options: UnpackOptions = {}): T
    builtin

/// Parse anything from a slice (most likely, you'll call it for structures).
/// \`\`\`
/// var msg = CounterIncrement.fromSlice(cs);
/// \`\`\`
/// All fields are read from a slice immediately unless \`lazy\`.
/// If a slice is corrupted, an exception is thrown (most likely, excode 9 "cell underflow").
/// Note, that a passed slice is NOT mutated, its internal pointer is NOT shifted.
/// If you need to mutate it, like \`cs.loadInt()\`, consider calling \`cs.loadAny<CounterIncrement>()\`.
@pure
fun T.fromSlice(rawSlice: slice, options: UnpackOptions = {}): T
    builtin

/// Parse anything from a slice, shifting its internal pointer.
/// Similar to \`slice.loadUint()\` and others, but allows loading structures.
/// \`\`\`
/// var st: MyStorage = cs.loadAny();     // or cs.loadAny<MyStorage>()
/// \`\`\`
/// Similar to \`MyStorage.fromSlice(cs)\`, but called as a slice method and mutates the slice.
/// Note: \`options.assertEndAfterReading\` is ignored by this function, because it's actually intended
/// to read data from the middle.
@pure
fun slice.loadAny<T>(mutate self, options: UnpackOptions = {}): T
    builtin

/// Skip anything in a slice, shifting its internal pointer.
/// Similar to \`slice.skipBits()\` and others, but allows skipping structures.
/// \`\`\`
/// struct TwoInts { a: int32; b: int32; }
/// cs.skipAny<TwoInts>();    // skips 64 bits
/// \`\`\`
@pure
fun slice.skipAny<T>(mutate self, options: UnpackOptions = {}): self
    builtin

/// Store anything to a builder.
/// Similar to \`builder.storeUint()\` and others, but allows storing structures.
/// \`\`\`
/// var b = beginCell().storeUint(32).storeAny(msgBody).endCell();
/// \`\`\`
@pure
fun builder.storeAny<T>(mutate self, v: T, options: PackOptions = {}): self
    builtin

/// Forces an object created by \`lazy\` to load fully. Returns the remaining slice (having read all fields).
/// Since \`options.assertEndAfterReading\` is ignored by \`lazy\` (fields are loaded on demand),
/// this method can help you overcome this, if you really need to check input consistency.
/// \`\`\`
/// val msg = lazy CounterMessage.fromSlice(s);
/// match (msg) {     // it's a lazy match, without creating a union on the stack
///     CounterIncrement => {
///         ...
///         newCounter = curCounter + msg.incBy;   // \`incBy\` loaded here, on demand
///         msg.forceLoadLazyObject().assertEnd()  // the purpose: get remainder
///     }
/// }
/// \`\`\`
/// Note: while [slice.assertEnd] may seem reasonable, these checks are avoided in practice,
/// because the purpose of \`lazy\` is to auto-detect and load only necessary fields, not up to the end.
@pure
fun T.forceLoadLazyObject(self): slice
    builtin

/// Parse data from already loaded cell reference.
/// \`\`\`
/// struct MyStorage { ... extra: Cell<ExtraData> }
///
/// var st = MyStorage.fromCell(contract.getData());
/// // st.extra is cell; if we need to unpack it, we do
/// var extra = st.extra.load();   // it's ExtraData, unpacked from loaded ref
/// \`\`\`
@pure
fun Cell<T>.load(self, options: UnpackOptions = {}): T
    builtin

/// Converts a typed cell into a slice.
@pure
fun Cell<T>.beginParse(self): slice
    asm "CTOS"

/// Returns hash of a typed cell, same as [cell.hash].
@pure
fun Cell<T>.hash(self): uint256
    asm "HASHCU"

/// \`RemainingBitsAndRefs\` is a special built-in type to get "all the rest" slice tail on reading.
/// \`\`\`
/// struct JettonMessage {
///     ... some fields
///     forwardPayload: RemainingBitsAndRefs;
/// }
/// \`\`\`
/// When you deserialize JettonMessage, forwardPayload contains "everything left after reading fields above".
type RemainingBitsAndRefs = slice

/// Creates a cell with zero bits and references.
/// Equivalent to \`beginCell().endCell()\`.
@pure
fun createEmptyCell(): cell
    asm "<b b> PUSHREF"

/// Creates a slice with zero remaining bits and references.
/// Equivalent to \`beginCell().endCell().beginParse()\`.
@pure
fun createEmptySlice(): slice
    asm "x{} PUSHSLICE"


/**
    Signature checks, hashing, cryptography.
 */

/// Compile-time function that calculates crc32 of a constant string.
/// \`\`\`
/// const op = stringCrc32("some_str") // = 4013618352 = 0xEF3AF4B0
/// \`\`\`
/// Note: \`stringCrc32(slice_var)\` does NOT work! It accepts a constant string and works at compile-time.
@pure
@deprecated("""use \`"str".crc32()\` instead of \`stringCrc32("str")\`""")
fun stringCrc32(constString: string): int
    builtin

/// Compile-time function that calculates crc16 (XMODEM) of a constant string.
/// \`\`\`
/// const op = stringCrc16("some_str") // = 53407 = 0xD09F
/// \`\`\`
/// Note: \`stringCrc16(slice_var)\` does NOT work! It accepts a constant string and works at compile-time.
@pure
@deprecated("""use \`"str".crc16()\` instead of \`stringCrc16("str")\`""")
fun stringCrc16(constString: string): int
    builtin

/// Compile-time function that calculates sha256 of a constant string and returns 256-bit integer.
/// \`\`\`
/// const hash = stringSha256("some_crypto_key")
/// \`\`\`
/// Note: it's a compile-time function, \`stringSha256(slice_var)\` does NOT work!
/// Use [slice.bitsHash] or [slice.hash] (declared below) to hash a slice without/with its refs at runtime.
@pure
@deprecated("""use \`"str".sha256()\` instead of \`stringSha256("str")\`""")
fun stringSha256(constString: string): int
    builtin

/// Compile-time function that calculates sha256 of a constant string and takes the first 32 bits.
/// \`\`\`
/// const minihash = stringSha256_32("some_crypto_key")
/// \`\`\`
/// Note: \`stringSha256_32(slice_var)\` does NOT work! It accepts a constant string and works at compile-time.
@pure
@deprecated("""use \`"str".sha256_32()\` instead of \`stringSha256_32("str")\`""")
fun stringSha256_32(constString: string): int
    builtin

/// Compile-time function that takes N-chars ascii string and interprets it as a number in base 256.
/// \`\`\`
/// const value = stringToBase256("AB") // = 16706 (65*256 + 66)
/// \`\`\`
/// Note: \`stringToBase256(slice_var)\` does NOT work! It accepts a constant string and works at compile-time.
@pure
@deprecated("""use \`"str".toBase256()\` instead of \`stringToBase256("str")\`""")
fun stringToBase256(constString: string): int
    builtin

/// Computes the representation hash of a cell and returns it as a 256-bit unsigned integer.
/// Useful for signing and checking signatures of arbitrary entities represented by a tree of cells.
@pure
fun cell.hash(self): uint256
    asm "HASHCU"

/// Computes the hash of a slice and returns it as a 256-bit unsigned integer.
/// The result is the same as [cell.hash] for a cell containing data and references from this slice.
@pure
fun slice.hash(self): uint256
    asm "NEWC" "STSLICE" "HASHBU"

/// Computes the hash of a builder and returns it as a 256-bit unsigned integer.
/// The same as "transform builder to cell + calc hash of that cell", but without cell creation.
@pure
fun builder.hash(self): uint256
    asm "HASHBU"

/// Computes sha256 of the data bits of a slice (without refs). 
/// If the bit length is not divisible by eight, throws a cell underflow exception.
@pure
fun slice.bitsHash(self): uint256
    asm "SHA256U"

/// Checks whether two cells are equal by comparing their hashes.
/// Used internally by the compiler for operators \`==\` and \`!=\` on cells.
@pure
fun cell.hashEqual(self, another: cell): bool
    asm "HASHCU" "SWAP" "HASHCU" "EQUAL"

/// Checks the Ed25519-\`signature\` of a \`hash\` (a 256-bit unsigned integer, usually computed as the hash of some data)
/// using \`publicKey\` (also represented by a 256-bit unsigned integer).
/// The signature must contain at least 512 data bits; only the first 512 bits are used.
/// Note that \`CHKSIGNU\` creates a 256-bit slice with the hash and calls \`CHKSIGNS\`.
/// That is, if parameter \`hash\` is computed as the hash of some data, these data are hashed twice,
/// the second hashing occurring inside \`CHKSIGNS\`.
@pure
fun isSignatureValid(hash: int, signature: slice, publicKey: int): bool
    asm "CHKSIGNU"

/// Checks whether \`signature\` is a valid Ed25519-signature of the data portion of slice \`data\` using \`publicKey\`,
/// similarly to [isSignatureValid].
/// If the bit length of \`data\` is not divisible by eight, throws a cell underflow exception.
/// The verification of Ed25519 signatures is the standard one,
/// with sha256 used to reduce \`data\` to the 256-bit number that is actually signed.
@pure
fun isSliceSignatureValid(data: slice, signature: slice, publicKey: int): bool
    asm "CHKSIGNS"

/// \`random\` is a built-in struct, it has only static methods.
/// Example: \`random.uint256()\` and other methods.
struct random

/// Generates a new pseudo-random unsigned 256-bit integer x.
/// Ensure you've called [random.initialize] in advance to make it unpredictable!
fun random.uint256(): uint256
    asm "RANDU256"

/// Generates a new pseudo-random integer z in the range 0..limit−1 (or limit..−1, if upto < 0).
/// More precisely, an unsigned random value x is generated as in random; then z := x * limit / 2^256 is computed.
/// Ensure you've called [random.initialize] in advance to make it unpredictable!
fun random.range(limit: int): int
    asm "RAND"

/// Returns the current random seed used to generate pseudo-random numbers.
@pure
fun random.getSeed(): uint256
    asm "RANDSEED"

/// Sets the random seed to the provided value.
fun random.setSeed(seed: uint256): void
    asm "SETRAND"

/// Initializes (mixes) random seed with the provided value.
fun random.initializeBy(mixSeedWith: uint256): void
    asm "ADDRAND"

/// Initializes random seed with current time to make random generation unpredictable.
/// Typically, you call this function once before calling [random.uint256] / [random.range].
fun random.initialize(): void
    asm "LTIME" "ADDRAND"


/**
    Size computation primitives.
    They may be useful for computing storage fees of user-provided data.
 */

/// Returns \`(x, y, z, -1)\` or \`(null, null, null, 0)\`.
/// Recursively computes the count of distinct cells \`x\`, data bits \`y\`, and cell references \`z\`
/// in a tree of cells, effectively returning the total storage used by this tree taking into account
/// the identification of equal cells.
/// The values of \`x\`, \`y\`, and \`z\` are computed by a depth-first traversal of this tree,
/// with a hash table of visited cell hashes used to prevent visits of already-visited cells.
/// The total count of visited cells \`x\` cannot exceed non-negative \`maxCells\`;
/// otherwise the computation is aborted before visiting the \`(maxCells + 1)\`-st cell and
/// a zero flag is returned to indicate failure.
@pure
fun cell.calculateSize(self, maxCells: int): (int, int, int, bool)
    asm "CDATASIZEQ NULLSWAPIFNOT2 NULLSWAPIFNOT"

/// Similar to [cell.calculateSize], but accepting a slice instead of a cell.
/// The returned value of \`x\` does not take into account the cell that contains the slice itself;
/// however, its data bits and cell references are accounted for in \`y\` and \`z\`.
@pure
fun slice.calculateSize(self, maxCells: int): (int, int, int, bool)
    asm "SDATASIZEQ NULLSWAPIFNOT2 NULLSWAPIFNOT"

/// A non-quiet version of [cell.calculateSize] that throws a cell overflow exception (\`8\`) on failure.
fun cell.calculateSizeStrict(self, maxCells: int): (int, int, int)
    asm "CDATASIZE"

/// A non-quiet version of [slice.calculateSize] that throws a cell overflow exception (\`8\`) on failure.
fun slice.calculateSizeStrict(self, maxCells: int): (int, int, int)
    asm "SDATASIZE"

/// Returns the depth of a cell: 0 if no references, otherwise 1 + maximum of depths of all references.
/// When called for \`null\`, returns 0.
@pure
fun cell?.depth(self): int
    asm "CDEPTH"

/// Returns the depth of a cell: 0 if no references, otherwise 1 + maximum of depths of all references.
@pure
fun slice.depth(self): int
    asm "SDEPTH"

/// Returns the depth of a builder: 0 if no references, otherwise 1 + maximum of depths of all references.
@pure
fun builder.depth(self): int
    asm "BDEPTH"


/**
    Debug primitives.
    Only works for local TVM execution with debug level verbosity.
 */

/// \`debug\` is a built-in struct, it has only static methods.
/// Example: \`debug.print(v)\` and other methods.
struct debug

/// Dump a variable to the debug log as a raw TVM stack value.
fun debug.print<T>(x: T): void
    builtin

/// Dump a string to the debug log.
fun debug.printString(x: string): void
    builtin

/// Dumps the stack (at most the top 255 values) and shows the total stack depth.
fun debug.dumpStack(): void
    builtin


/*
    Implementation of \`map<K, V>\` and methods over it.

    Note: maps in TVM are called "dictionaries". They are stored as nested cells 
  (each value in a separate cell, probably with intermediate cells-nodes).
  Constructing even small dicts is gas-expensive, because cell creation costs a lot.

    \`map<K, V>\` is a "human-readable interface" over low-level dictionaries. The compiler,
  knowing types of K and V, effectively generates asm instructions for storing/loading values.
  But still, of course, it's a tree of cells, it's just a TVM dictionary.
 */

/// Returns an empty typed map. It's essentially "PUSHNULL", since TVM NULL represents an empty map.
/// \`\`\`
/// // old way, deprecated:
/// var m: map<int8, int32> = createEmptyMap();
///
/// // new way, preferred:
/// var m: map<int8, int32> = [];
/// \`\`\`
@pure
@deprecated("use \`[]\` instead of \`createEmptyMap\`")
fun createEmptyMap<K, V>(): map<K, V>
    builtin

/// Converts a low-level TVM dictionary to a typed map.
/// Actually, does nothing: accepts an "optional cell" and returns the same "optional cell",
/// so if you specify key/value types incorrectly, it will fail later, at \`map.get\` and similar.
@pure
fun createMapFromLowLevelDict<K, V>(d: dict): map<K, V>
    builtin

/// Converts a high-level map to a low-level TVM dictionary.
/// Actually, does nothing: returns the same "optional cell".
@pure
fun map<K, V>.toLowLevelDict(self): dict
    asm "NOP"

/// Checks whether a map is empty (whether a cell is null).
/// 
/// Note: a check \`m == null\` will not work, use \`m.isEmpty()\`.
@pure
fun map<K, V>.isEmpty(self): bool
    asm "DICTEMPTY"

/// Checks whether a key exists in a map.
@pure
fun map<K, V>.exists(self, key: K): bool
    builtin

/// Gets an element by key. If not found, does NOT throw, just returns isFound = false.
/// \`\`\`
/// val r = m.get(123);
/// if (r.isFound) {
///     r.loadValue()
/// }
/// \`\`\`
@pure
fun map<K, V>.get(self, key: K): MapLookupResult<V>
    builtin

/// Gets an element by key and throws if it doesn't exist.
@pure
fun map<K, V>.mustGet(self, key: K, throwIfNotFound: int = 9): V
    builtin

/// Sets an element by key.
/// \`\`\`
/// m.set(k, 3);
/// \`\`\`
/// Since it returns \`self\`, calls may be chained.
@pure
fun map<K, V>.set(mutate self, key: K, value: V): self
    builtin

/// Sets an element and returns the previous element at that key. If no previous, isFound = false.
/// \`\`\`
/// val prev = m.setAndGetPrevious(k, 3);
/// if (prev.isFound) {
///     prev.loadValue()
/// }
/// \`\`\`
@pure
fun map<K, V>.setAndGetPrevious(mutate self, key: K, value: V): MapLookupResult<V>
    builtin

/// Sets an element only if the key already exists. Returns whether an element was replaced.
@pure
fun map<K, V>.replaceIfExists(mutate self, key: K, value: V): bool
    builtin

/// Sets an element only if the key already exists and returns the previous element at that key.
@pure
fun map<K, V>.replaceAndGetPrevious(mutate self, key: K, value: V): MapLookupResult<V>
    builtin

/// Sets an element only if the key does not exist. Returns whether an element was added.
@pure
fun map<K, V>.addIfNotExists(mutate self, key: K, value: V): bool
    builtin

/// Sets an element only if the key does not exist. If exists, returns an old value.
@pure
fun map<K, V>.addOrGetExisting(mutate self, key: K, value: V): MapLookupResult<V>
    builtin

/// Delete an element at the key. Returns whether an element was deleted. 
@pure
fun map<K, V>.delete(mutate self, key: K): bool
    builtin

/// Delete an element at the key and returns the deleted element. If not exists, isFound = false.
/// \`\`\`
/// val prev = m.deleteAndGetDeleted(k);
/// if (prev.isFound) {
///     prev.loadValue()
/// }
/// \`\`\`
@pure
fun map<K, V>.deleteAndGetDeleted(mutate self, key: K): MapLookupResult<V>
    builtin

/// Finds the first (minimal) element in a map. If key are integers, it's the minimal integer.
/// If keys are addresses or complex structures (represented as slices), it's lexicographically smallest.
/// For an empty map, just returns isFound = false.
/// Useful for iterating over a map:
/// \`\`\`
/// var r = m.findFirst();
/// while (r.isFound) {
///     // ... use r.getKey() and r.loadValue()
///     r = m.iterateNext(r)
/// }
/// \`\`\`
@pure
fun map<K, V>.findFirst(self): MapEntry<K, V>
    builtin

/// Finds the last (maximal) element in a map. If key are integers, it's the maximal integer.
/// If keys are addresses or complex structures (represented as slices), it's lexicographically largest.
/// For an empty map, just returns isFound = false.
/// Useful for iterating over a map:
/// \`\`\`
/// var r = m.findLast();
/// while (r.isFound) {
///     // ... use r.getKey() and r.loadValue()
///     r = m.iteratePrev(r)
/// }
/// \`\`\`
@pure
fun map<K, V>.findLast(self): MapEntry<K, V>
    builtin

/// Finds an element with key > pivotKey.
/// Don't forget to check \`isFound\` before using \`getKey()\` and \`loadValue()\` of the result.
@pure
fun map<K, V>.findKeyGreater(self, pivotKey: K): MapEntry<K, V>
    builtin

/// Finds an element with key >= pivotKey.
/// Don't forget to check \`isFound\` before using \`getKey()\` and \`loadValue()\` of the result.
@pure
fun map<K, V>.findKeyGreaterOrEqual(self, pivotKey: K): MapEntry<K, V>
    builtin

/// Finds an element with key < pivotKey.
/// Don't forget to check \`isFound\` before using \`getKey()\` and \`loadValue()\` of the result.
@pure
fun map<K, V>.findKeyLess(self, pivotKey: K): MapEntry<K, V>
    builtin

/// Finds an element with key <= pivotKey.
/// Don't forget to check \`isFound\` before using \`getKey()\` and \`loadValue()\` of the result.
@pure
fun map<K, V>.findKeyLessOrEqual(self, pivotKey: K): MapEntry<K, V>
    builtin

/// Iterate over a map in ascending order.
/// \`\`\`
/// // iterate for all keys >= 10 up to the end
/// var r = m.findKeyGreaterOrEqual(10);
/// while (r.isFound) {
///     // ... use r.getKey() and r.loadValue()
///     r = m.iterateNext(r)
/// }
/// \`\`\`
@pure
fun map<K, V>.iterateNext(self, current: MapEntry<K, V>): MapEntry<K, V>
    builtin

/// Iterate over a map in reverse order.
/// \`\`\`
/// // iterate for all keys < 10 down lo lowest
/// var r = m.findKeyLess(10);
/// while (r.isFound) {
///     // ... use r.getKey() and r.loadValue()
///     r = m.iteratePrev(r)
/// }
/// \`\`\`
@pure
fun map<K, V>.iteratePrev(self, current: MapEntry<K, V>): MapEntry<K, V>
    builtin

/// \`MapLookupResult\` is a return value of \`map.get\`, \`map.setAndGetPrevious\`, and similar.
/// Instead of returning a nullable value (that you'd check on null before usage),
/// this struct is returned (and you check \`isFound\` before usage).
/// \`\`\`
/// val r = m.get(key);
/// if (r.isFound) {
///     r.loadValue()    // unpacks a returned slice
/// }
/// \`\`\`
struct MapLookupResult<TValue> {
    private readonly rawSlice: slice?        // holds encoded value, present if isFound
    isFound: bool
}

@pure
fun MapLookupResult<TValue>.loadValue(self): TValue {
    // it's assumed that you check for \`isFound\` before calling \`loadValue()\`;
    // note that \`assertEnd\` is called to ensure no remaining data left besides TValue
    return TValue.fromSlice(self.rawSlice!, {
        assertEndAfterReading: true
    })
}

@pure
fun MapLookupResult<slice>.loadValue(self): slice {
    return self.rawSlice!
}

/// \`MapEntry\` is a return value of \`map.findFirst\`, \`map.iterateNext\`, and similar.
/// You should check for \`isFound\` before calling \`getKey()\` and \`loadValue()\`.
/// \`\`\`
/// var r = m.findFirst();
/// while (r.isFound) {
///     // ... use r.getKey() and r.loadValue()
///     r = m.iterateNext(r)
/// }
/// \`\`\`
struct MapEntry<K, V> {
    private readonly rawValue: slice?        // holds encoded value, present if isFound
    private readonly key: K
    isFound: bool
}

@pure
fun MapEntry<K, V>.getKey(self): K {
    // it's assumed that you check for \`isFound\` before calling \`getKey()\`;
    // (otherwise, it will contain an incorrect stack slot with \`null\` value, not K)
    return self.key
}

@pure
fun MapEntry<K, V>.loadValue(self): V {
    // it's assumed that you check for \`isFound\` before calling \`loadValue()\`;
    // note that \`assertEnd\` is inserted to ensure no remaining data left besides TValue
    return V.fromSlice(self.rawValue!, {
        assertEndAfterReading: true
    })
}

@pure
fun MapEntry<K, slice>.loadValue(self): slice {
    return self.rawValue!
}


/**
    Strings.
    A string is a language primitive — a cell containing snake-formatted text data.
    See [string] built-in type.
    Also see \`@stdlib/strings.tolk\` — additional methods for strings (including \`StringBuilder\`),
  which must be explicitly imported to be used.
 */

/// Begins parsing the string content. Use for accessing raw bytes.
/// Remember that a string may be a snake: not just "ab", but "a" + (ref "b").
@pure
fun string.beginParse(self): slice
    asm "CTOS"

/// Compile-time method that returns raw bytes of a short string literal as a slice.
/// Use this when you need a slice constant, not a string.
/// \`\`\`
/// const s: slice = "hello".literalSlice()
/// \`\`\`
/// Unlike \`beginParse()\`, it works at compile-time and is allowed in constant expressions.
@pure
fun string.literalSlice(self): slice
    builtin

/// Compile-time method that calculates crc32 of a constant string.
/// \`\`\`
/// const op = "some_str".crc32()  // = 4013618352 = 0xEF3AF4B0
/// \`\`\`
/// Note: this is a compile-time method, it works only with string literals.
@pure
fun string.crc32(self): int
    builtin

/// Compile-time method that calculates crc16 (XMODEM) of a constant string.
/// \`\`\`
/// const op = "some_str".crc16()  // = 53407 = 0xD09F
/// \`\`\`
/// Note: this is a compile-time method, it works only with string literals.
@pure
fun string.crc16(self): int
    builtin

/// Compile-time method that calculates sha256 of a constant string.
/// \`\`\`
/// const hash = "some_crypto_key".sha256()
/// \`\`\`
/// Note: this is a compile-time method, it works only with string literals.
@pure
fun string.sha256(self): int
    builtin

/// Compile-time method that calculates sha256 of a constant string and takes the first 32 bits.
/// \`\`\`
/// const minihash = "some_crypto_key".sha256_32()
/// \`\`\`
/// Note: this is a compile-time method, it works only with string literals.
@pure
fun string.sha256_32(self): int
    builtin

/// Compile-time method that converts a constant hex-encoded string to a slice.
/// \`\`\`
/// const v = "abcdef".hexToSlice()  // = slice with 3 bytes: 0xAB,0xCD,0xEF
/// \`\`\`
/// Note: this is a compile-time method, it works only with string literals.
@pure
fun string.hexToSlice(self): slice
    builtin

/// Compile-time method that takes N-chars ascii string and interprets it as a number in base 256.
/// \`\`\`
/// const value = "AB".toBase256() // = 16706 (65*256 + 66)
/// \`\`\`
/// Note: this is a compile-time method, it works only with string literals.
@pure
fun string.toBase256(self): int
    builtin


/**
    Slice primitives: parsing cells.
    When you _load_ some data, you mutate the slice (shifting an internal pointer on the stack).
    When you _preload_ some data, you just get the result without mutating the slice.
 */

/// Compile-time function that converts a constant hex-encoded string to N/2 bytes.
/// \`\`\`
/// const v = stringHexToSlice("abcdef") // = slice with 3 bytes: 0xAB,0xCD,0xEF
/// \`\`\`
/// Note: \`stringHexToSlice(slice_var)\` does NOT work! It accepts a constant string and works at compile-time.
@pure
@deprecated("""use \`"hex".hexToSlice()\` instead of \`stringHexToSlice("hex")\`""")
fun stringHexToSlice(constStringBytesHex: string): slice
    builtin

/// Converts a cell into a slice.
@pure
fun cell.beginParse(self): slice
    asm "CTOS"

/// Checks if a slice is empty. If not, throws an exception with code 9.
fun slice.assertEnd(self): void
    asm "ENDS"

/// Loads the next reference from a slice.
@pure
fun slice.loadRef(mutate self): cell
    asm( -> 1 0) "LDREF"

/// Preloads the next reference from a slice.
@pure
fun slice.preloadRef(self): cell
    asm "PLDREF"

/// Loads a signed len-bit integer from a slice.
@pure
fun slice.loadInt(mutate self, len: int): int
    builtin

/// Loads an unsigned len-bit integer from a slice.
@pure
fun slice.loadUint(mutate self, len: int): int
    builtin

/// Loads the first \`0 ≤ len ≤ 1023\` bits from a slice.
@pure
fun slice.loadBits(mutate self, len: int): slice
    builtin

/// Preloads a signed len-bit integer from a slice.
@pure
fun slice.preloadInt(self, len: int): int
    builtin

/// Preloads an unsigned len-bit integer from a slice.
@pure
fun slice.preloadUint(self, len: int): int
    builtin

/// Preloads the first \`0 ≤ len ≤ 1023\` bits from a slice.
@pure
fun slice.preloadBits(self, len: int): slice
    builtin

/// Loads the serialized amount of nanotoncoins (any unsigned integer up to \`2^120 - 1\`).
@pure
fun slice.loadCoins(mutate self): coins
    asm( -> 1 0) "LDGRAMS"

/// Loads a boolean from a slice.
@pure
fun slice.loadBool(mutate self): bool
    asm( -> 1 0) "1 LDI"

/// Shifts a slice pointer to len bits forward.
@pure
fun slice.skipBits(mutate self, len: int): self
    builtin

/// Returns the first \`0 ≤ len ≤ 1023\` bits of a slice.
@pure
fun slice.getFirstBits(self, len: int): slice
    asm "SDCUTFIRST"

/// Returns all but the last \`0 ≤ len ≤ 1023\` bits of a slice.
@pure
fun slice.removeLastBits(mutate self, len: int): self
    asm "SDSKIPLAST"

/// Returns the last \`0 ≤ len ≤ 1023\` bits of a slice.
@pure
fun slice.getLastBits(self, len: int): slice
    asm "SDCUTLAST"

/// Returns \`0 ≤ len ≤ 1023\` bits of a slice starting from \`0 ≤ offset ≤ 1023\`.
/// (in other words, extracts a bit substring \`[offset, len)\` out of the slice)
@pure
fun slice.getMiddleBits(self, offset: int, len: int): slice
    asm "SDSUBSTR"

/// Loads a dictionary (TL HashMapE structure, represented as TVM cell) from a slice.
@pure
fun slice.loadDict(mutate self): dict
    asm( -> 1 0) "LDDICT"

/// Preloads a dictionary (cell) from a slice.
@pure
fun slice.preloadDict(self): dict
    asm "PLDDICT"

/// Skip a dictionary from a slice.
@pure
fun slice.skipDict(mutate self): self
    asm "SKIPDICT"

/// Loads (Maybe ^Cell) from a slice.
/// In other words, loads 1 bit: if it's true, loads the first ref, otherwise returns \`null\`.
@pure
fun slice.loadMaybeRef(mutate self): cell?
    asm( -> 1 0) "LDOPTREF"

/// Preloads (Maybe ^Cell) from a slice.
@pure
fun slice.preloadMaybeRef(self): cell?
    asm "PLDOPTREF"

/// Skips (Maybe ^Cell) in a slice.
@pure
fun slice.skipMaybeRef(mutate self): self
    asm "SKIPOPTREF"

/// Loads a string from a slice. A string is a nested ref, so just load a ref.
@pure
fun slice.loadString(mutate self): string
    asm( -> 1 0) "LDREF"


/**
    Builder primitives: constructing cells.
    When you _store_ some data, you mutate the builder (shifting an internal pointer on the stack).
    All the primitives below first check whether there is enough space in the \`builder\`,
    and only then check the range of the value being serialized.
 */

/// Creates a new empty builder.
@pure
fun beginCell(): builder
    asm "NEWC"

/// Converts a builder into an ordinary cell.
@pure
fun builder.endCell(self): cell
    asm "ENDC"

/// Converts a builder into a slice.
/// The same as \`b.endCell().beginParse()\`
@pure
fun builder.toSlice(self): slice
    asm "BTOS"

/// Stores a reference to a cell into a builder.
@pure
fun builder.storeRef(mutate self, c: cell): self
    asm(c self) "STREF"

/// Stores a signed len-bit integer into a builder (\`0 ≤ len ≤ 257\`).
@pure
fun builder.storeInt(mutate self, x: int, len: int): self
    builtin

/// Stores an unsigned len-bit integer into a builder (\`0 ≤ len ≤ 256\`).
@pure
fun builder.storeUint(mutate self, x: int, len: int): self
    builtin

/// Stores a slice into a builder.
@pure
fun builder.storeSlice(mutate self, s: slice): self
    asm(s self) "STSLICE"

/// Stores a string into a builder (it's a TVM cell, stored as a ref).
@pure
fun builder.storeString(mutate self, s: string): self
    asm(s self) "STREF"

/// Stores an internal address into a builder.
@pure
fun builder.storeAddress(mutate self, addr: address): self
    asm(addr self) "STSTDADDR"

/// Stores an internal address or '00' if the parameter is \`null\`.
/// '00' (two zero bits) is a standard representation of a "missing" ("none") address.
@pure
fun builder.storeAddressOpt(mutate self, addrOrNull: address?): self
    asm(addrOrNull self) "STOPTSTDADDR"

/// Stores "any address" (internal/external/none) into a builder.
@pure
fun builder.storeAddressAny(mutate self, addrAny: any_address): self
    asm(addrAny self) "STSLICE"

/// Stores a "none address": '00' (two zero bits) is TL addr_none$00.
@pure
fun builder.storeAddressNone(mutate self): self
    asm "b{00} STSLICECONST"

/// Stores the amount of nanotoncoins into a builder.
@pure
fun builder.storeCoins(mutate self, x: coins): self
    builtin

/// Stores a boolean into a builder.
@pure
fun builder.storeBool(mutate self, x: bool): self
    builtin

/// Stores a low-level TVM dictionary (optional cell) into a builder.
/// In other words, if \`c\` is null, stores \`0\`; if \`c\` is not null, stores \`1\` and a reference to \`c\`.
@pure
fun builder.storeDict(mutate self, c: dict): self
    asm(c self) "STDICT"

/// Stores (Maybe ^Cell) into a builder.
/// In other words, if \`c\` is null, stores \`0\`; if \`c\` is not null, stores \`1\` and a reference to \`c\`.
@pure
fun builder.storeMaybeRef(mutate self, c: cell?): self
    asm(c self) "STOPTREF"

/// Concatenates two builders.
@pure
fun builder.storeBuilder(mutate self, from: builder): self
    asm(from self) "STB"


/**
    Slice size primitives.
 */

/// Returns the number of references in a slice.
@pure
fun slice.remainingRefsCount(self): int
    asm "SREFS"

/// Returns the number of data bits in a slice.
@pure
fun slice.remainingBitsCount(self): int
    asm "SBITS"

/// Returns both the number of data bits and the number of references in a slice.
@pure
fun slice.remainingBitsAndRefsCount(self): (int, int)
    asm "SBITREFS"

/// Checks whether a slice is empty (i.e., contains no bits of data and no cell references).
@pure
fun slice.isEmpty(self): bool
    asm "SEMPTY"

/// Checks whether a slice has no bits of data.
@pure
fun slice.isEndOfBits(self): bool
    asm "SDEMPTY"

/// Checks whether a slice has no references.
@pure
fun slice.isEndOfRefs(self): bool
    asm "SREMPTY"

/// Checks whether data parts of two slices coinside.
@pure
fun slice.bitsEqual(self, b: slice): bool
    asm "SDEQ"

/// Returns the number of cell references already stored in a builder.
@pure
fun builder.refsCount(self): int
    asm "BREFS"

/// Returns the number of data bits already stored in a builder.
@pure
fun builder.bitsCount(self): int
    asm "BBITS"


/**
    Address manipulation primitives.
 */

/// Compile-time function that parses a valid contract address.
/// \`\`\`
/// address("EQCRDM9h4k3UJdOePPuyX40mCgA4vxge5Dc5vjBR8djbEKC5")
/// address("0:527964d55cfa6eb731f4bfc07e9d025098097ef8505519e853986279bd8400d8")
/// \`\`\`
/// Returns \`address\`, which can be stored in a struct, compared with \`==\`, etc.
@pure
fun address(stdAddress: string): address
    builtin

/// Extracts workchain and hash from a standard (internal) address.
@pure
fun address.getWorkchainAndHash(self): (int8, uint256)
    asm "REWRITESTDADDR"

/// Extracts workchain from a standard (internal) address.
@pure
fun address.getWorkchain(self): int8
    asm "REWRITESTDADDR" "DROP"

/// Creates a slice representing "none external address" (TL addr_none$00 — two zero bits).
/// Note! Use this in combination with EXTERNAL addresses (e.g. for [createExternalLogMessage]).
/// To represent "internal or none" address (which you most likely need to), 
/// use \`address?\` (nullable) and \`null\` accordingly.
@pure
fun createAddressNone(): any_address
    asm "b{00} PUSHSLICE"

/// Returns if "any address" is "none".
/// In TL/B, it's addr_none$00.
@pure
fun any_address.isNone(self): bool
    asm "b{00} SDBEGINSQ" "NIP"

/// Returns if "any address" is a standard (internal) address: workchain + hash.
/// Then it can be safely cast to \`address\`. 
/// In TL/B, it's addr_std$10.
@pure
fun any_address.isInternal(self): bool
    asm "b{10} SDBEGINSQ" "NIP"

/// Returns if "any address" is an external address, used to communication with the outside world.
/// In TL/B, it's addr_extern$01.
@pure
fun any_address.isExternal(self): bool
    asm "b{01} SDBEGINSQ" "NIP"

/// Extracts workchain and hash from a standard (internal) address.
/// If the address is not internal, throws a cell deserialization exception.
@pure
fun any_address.getWorkchainAndHash(self): (int8, uint256)
    asm "REWRITESTDADDR"

/// Extracts workchain from a standard (internal) address.
/// If the address is not internal, throws a cell deserialization exception.
@pure
fun any_address.getWorkchain(self): int8
    asm "REWRITESTDADDR" "DROP"

/// Casts \`any_address\` to \`address\` checking that "any" is actually "internal".
/// E.g., \`addr = myAny.castToInternal()\` does the check and returns \`address\`.
/// To skip the runtime check, use unsafe type casting instead: \`addr = myAny as address\`.
@pure
fun any_address.castToInternal(self): address
    asm "DUP" "b{10} SDBEGINSQ" "NIP" "9 THROWIFNOT"

/// Checks whether two addresses are equal. Equivalent to \`a == b\`.
/// Deprecated! Left for smoother transition from FunC, where you used \`slice\` everywhere.
/// Use just \`a == b\` and \`a != b\` to compare addresses, don't use bitsEqual.
@pure
@deprecated("use \`senderAddress == ownerAddress\`, not \`senderAddress.bitsEqual(ownerAddress)\`")
fun address.bitsEqual(self, b: address): bool
    asm "SDEQ"

/// Loads a standard (internal) address (containing workchain + hash). In TL/B, it's \`MsgAddressInt\`.
/// If an address is incorrect or is not internal, throws a deserialization exception (code 9).
@pure
fun slice.loadAddress(mutate self): address
    asm( -> 1 0) "LDSTDADDR"

/// Loads a standard (internal) address or returns \`null\` if the input is "none address".
/// If an address is incorrect or is not internal/none, throws a deserialization exception (code 9).
/// '00' (two zero bits) is a standard representation of a "missing" ("none") address.
/// As a result, you get either \`address\` (for internal) or \`null\` (for none).
@pure
fun slice.loadAddressOpt(mutate self): address?
    asm( -> 1 0) "LDOPTSTDADDR"

/// Loads a valid "any address" (internal/external/none). In TL/B, it's \`MsgAddress\`.
@pure
fun slice.loadAddressAny(mutate self): any_address
    asm( -> 1 0) "LDMSGADDR"


/**
    Reserving Toncoins on balance and its flags.
 */

/// mode = 0: Reserve exact amount of nanotoncoins
const RESERVE_MODE_EXACT_AMOUNT = 0
/// +1: Actually reserves all but amount, meaning \`currentContractBalance - amount\`
const RESERVE_MODE_ALL_BUT_AMOUNT = 1
/// +2: Actually set \`min(amount, currentContractBalance)\` (without this mode, if amount is greater, the action will fail)
const RESERVE_MODE_AT_MOST = 2
/// +4: [amount] is increased by the _original_ balance of the current account (before the compute phase).
const RESERVE_MODE_INCREASE_BY_ORIGINAL_BALANCE = 4
/// +8: Actually sets \`amount = -amount\` before performing any further actions.
const RESERVE_MODE_NEGATE_AMOUNT = 8
/// +16: If this action fails, the transaction will be bounced.
const RESERVE_MODE_BOUNCE_ON_ACTION_FAIL = 16

/// Creates an output action which would reserve Toncoins on balance.
/// For \`reserveMode\` consider constants above.
fun reserveToncoinsOnBalance(nanoTonCoins: coins, reserveMode: int): void
    asm "RAWRESERVE"

/// Similar to [reserveToncoinsOnBalance], but also accepts a dictionary extraAmount (represented by a cell or null)
/// with extra currencies. In this way currencies other than Toncoin can be reserved.
fun reserveExtraCurrenciesOnBalance(nanoTonCoins: coins, extraAmount: dict, reserveMode: int): void
    asm "RAWRESERVEX"


/**
    Creating and sending messages.
    Basic scenario:
    \`\`\`
    val outMsg = createMessage({ ... options });   // you get OutMessage
    outMsg.send(mode);
    \`\`\`
 */

/// mode = 0 is used for ordinary messages; the gas fees are deducted from the senging amount; action phaes should NOT be ignored.
const SEND_MODE_REGULAR = 0
/// +1 means that the sender wants to pay transfer fees separately.
const SEND_MODE_PAY_FEES_SEPARATELY = 1
/// +2 means that any errors arising while processing this message during the action phase should be ignored.
const SEND_MODE_IGNORE_ERRORS = 2
/// in the case of action fail - bounce transaction. No effect if SEND_MODE_IGNORE_ERRORS (+2) is used. [TVM UPGRADE 2023-07](https://docs.ton.org/learn/tvm-instructions/tvm-upgrade-2023-07#sending-messages)
const SEND_MODE_BOUNCE_ON_ACTION_FAIL = 16
/// mode = 32 means that the current account must be destroyed if its resulting balance is zero.
const SEND_MODE_DESTROY = 32
/// mode = 64 is used for messages that carry all the remaining value of the inbound message in addition to the value initially indicated in the new message.
const SEND_MODE_CARRY_ALL_REMAINING_MESSAGE_VALUE = 64
/// mode = 128 is used for messages that are to carry all the remaining balance of the current smart contract (instead of the value originally indicated in the message).
const SEND_MODE_CARRY_ALL_BALANCE = 128
/// do not create an action, only estimate fee. [TVM UPGRADE 2023-07](https://docs.ton.org/learn/tvm-instructions/tvm-upgrade-2023-07#sending-messages)
const SEND_MODE_ESTIMATE_FEE_ONLY = 1024
/// Other modes affect the fee calculation as follows:
/// +64 substitutes the entire balance of the incoming message as an outcoming value (slightly inaccurate, gas expenses that cannot be estimated before the computation is completed are not taken into account).
/// +128 substitutes the value of the entire balance of the contract before the start of the computation phase (slightly inaccurate, since gas expenses that cannot be estimated before the completion of the computation phase are not taken into account).

/// \`ExtraCurrenciesMap\` represents a dictionary of "extra currencies" other than TON coin.
/// They can be attached to a message (incoming and outgoing) and stored on contract's balance.
/// It's a dictionary \`[currencyID => amount]\` (amount is like \`coins\`, but encoded with a high precision). 
type ExtraCurrenciesMap = map<int32, varuint32>

/// \`ContractState\` is "code + data" of a contract.
/// Used in outgoing messages (StateInit) to initialize a destination contract.
struct ContractState {
    code: cell
    data: cell
}

/// \`AddressShardingOptions\` provides settings to calculate an address in another shard.
/// Consider [createMessage] and [address.calculateSameAddressInAnotherShard] for usage.
struct AddressShardingOptions {
    fixedPrefixLength: uint5        // shard depth, formerly splitDepth
    closeTo: address
}

/// \`AutoDeployAddress\` is a destination that initializes a receiver contract if it does not exist yet.
/// In order to do this, it contains StateInit — and calculates an address by a hash of StateInit.
/// \`\`\`
/// createMessage({
///     dest: {
///         workchain: 0,
///         stateInit: {
///             code: jettonWalletCode,
///             data: jettonWalletEmptyStorage.toCell()
///         }
///     },
///     ...
/// \`\`\`
/// You just provide code+data, and the compiler automatically calculates the destination address,
/// because in TON, the address of a contract, by definition, is a hash of its initial state.
/// You can also use it without sending a message. 
/// See [AutoDeployAddress.calculateAddress].
struct AutoDeployAddress {
    workchain: int8 = BASECHAIN
    stateInit: ContractState | cell
    toShard: AddressShardingOptions? = null
}

/// Deprecated, use [AutoDeployAddress.calculateAddress] instead.
/// Since late 2025 (TVM 12 has a cheap "builder-to-slice" instruction),
/// there is no need to use this function.
/// Historically its purpose was to avoid creating a cell and to return \`builder\` with a valid address.
@pure
@deprecated("use calculateAddress instead")
fun AutoDeployAddress.buildAddress(self): builder
    builtin

/// Constructs an address that a deployed contract will have.
/// For example, from a jetton minter, you want to calculate an address of a jetton wallet:
/// \`\`\`
/// val jwDeployed = calcDeployedJettonWallet(...);
/// val jwAddress = jwDeployed.calculateAddress();
/// \`\`\`
/// Just instead of \`dest\` for [createMessage], you use it is to calculate a jetton/nft address.
@pure
fun AutoDeployAddress.calculateAddress(self): address
    builtin

/// Checks that an address matches a deployed contract.
/// For example, from a jetton minter, you're checking whether a message is from a jetton wallet:
/// \`\`\`
/// val jwDeployed = calcDeployedJettonWallet(...);
/// val fromJW = jwDeployed.addressMatches(senderAddress);
/// \`\`\`
/// Just instead of \`dest\` for [createMessage], you use it is to check whether a sender is a jetton/nft.
/// Deprecated since late 2025 (TVM 12 with a cheap "builder-to-slice" instruction), to be removed later.
@pure
@deprecated("Use \`a == jwDeployed.calculateAddress()\` instead of \`jwDeployed.addressMatches(a)")
fun AutoDeployAddress.addressMatches(self, addr: address): bool
    builtin

/// \`BounceMode\` represents behavior of an outgoing message when a receiver failed to handle it.
/// If you specify to bounce a message, don't forget about the \`onBouncedMessage\` entrypoint.
/// \`\`\`
/// val outMsg = createMessage({
///     bounce: BounceMode.Only256BitsOfBody,
///     ...
/// });
/// 
/// fun onBouncedMessage(in: InMessageBounced) {
///     // in.bouncedBody is populated depending on BounceMode
///     // for instance, "Only256BitsOfBody" means 
///     // "bounced prefix 0xFFFFFFFF" and 256 bits of data
///     in.bouncedBody.skipBouncedPrefix();
///     val originalOpcode = in.bouncedBody.loadUint(32);
///     ...
/// }
/// \`\`\`
enum BounceMode {
    NoBounce                // a message will just disappear on error
    Only256BitsOfBody       // \`in.bouncedBody\` will be "0xFFFFFFFF" and the first 256 bits of outgoing body (most cheap)
    RichBounce              // \`in.bouncedBody\` will be struct RichBounceBody (most expensive, but allows accessing all data sent)
    RichBounceOnlyRootCell  // \`in.bouncedBody\` will be struct RichBounceBody without refs in \`originalBody\`
}

/// Before TVM-12, \`bounce\` was only true or false.
/// Now, use \`enum BounceMode\` instead. "true" behaves like \`Only256BitsOfBody\`.
@deprecated("use \`BounceMode.NoBounce\` instead of \`false\`")
type OldBounceMode = bool

/// Options for creating an outgoing message.
/// Consider [createMessage] for examples.
struct CreateMessageOptions<TBody = void> {
    /// whether a message will bounce back on error, and what piece of outgoing \`body\` it will contain
    bounce: BounceMode | OldBounceMode
    /// message value: attached tons (or tons + extra currencies)
    value: coins | (coins, ExtraCurrenciesMap)
    /// destination is either a provided address, or is auto-calculated by stateInit
    dest: address               // either just send a message to some address
        | builder               // ... or a manually constructed builder with a valid address
        | (int8, uint256)       // ... or to workchain + hash (also known as accountID)
        | AutoDeployAddress     // ... or "send to stateInit" aka deploy (address auto-calculated)
    /// body is any serializable object (or just miss this field for empty body)
    body: TBody
}

/// Creates a message (\`OutMessage\`) — a well-formatted message cell.
/// Typically, you just send it. In advanced scenarios, you can estimate fees or even postpone sending.
/// \`\`\`
/// val reply = createMessage({
///     bounce: BounceMode.NoBounce,
///     value: ton("0.05"),
///     dest: senderAddress,
///     body: RequestedInfo { ... }    // note: no toCell! just pass an object
/// });
/// reply.send(SEND_MODE_REGULAR);
/// \`\`\`
///
/// Hint: don't call \`body.toCell()\`, pass \`body\` directly!
///       (if body is small, it will be inlined without an expensive cell creation)
///       (if body is large, the compiler will automatically wrap it into a cell)
/// If you need an empty body, just miss the field \`body\`, fill other 3 fields.
@pure
fun createMessage<TBody>(options: CreateMessageOptions<TBody>): OutMessage
    builtin


/// \`ExtOutLogBucket\` is a variant of a custom external address for emitting logs "to the outer world".
/// It includes some "topic" (arbitrary number), that determines the format of the message body.
/// For example, you emit "deposit event" (reserving topic "deposit" = 123):
/// \`\`\`
/// dest: ExtOutLogBucket { topic: 123 }, 
/// body: DepositData { ... }
/// \`\`\`
/// and external indexers can index your emitted logs by destination address without parsing body.
/// Serialization details: '01' (addr_extern) + 256 (len) + 0x00 (prefix) + 248 bits = 267 in total
struct (0x00) ExtOutLogBucket {
    topic: uint248 | bits248
}

/// Options for creating an external outgoing message.
/// Consider [createExternalLogMessage] for examples.
struct CreateExternalLogMessageOptions<TBody = void> {
    /// destination is either an external address or a pattern to calculate it
    dest: any_address       // either some valid external/none address (not internal!)
        | builder           // ... or a manually constructed builder with a valid external address
        | ExtOutLogBucket   // ... or encode topic/eventID in destination
    /// body is any serializable object (or just miss this field for empty body)
    body: TBody
}

/// Creates an external message (\`OutMessage\`) — a well-formatted message cell.
/// Typically, you just send it. In advanced scenarios, you can estimate fees or even postpone sending.
/// \`\`\`
/// val emitMsg = createExternalLogMessage({
///     dest: createAddressNone(),
///     body: DepositEvent { ... }        // note: no \`toCell\`! just pass an object
/// });
/// emitMsg.send(SEND_MODE_REGULAR);
/// \`\`\`
/// Note: \`createMessage\` also returns \`OutMessage\`, it's okay: a composed message cell is universal.
/// Hint: don't call \`body.toCell()\`, pass \`body\` directly!
///       (if body is small, it will be inlined without an expensive cell creation)
///       (if body is large, the compiler will automatically wrap it into a cell)
@pure
fun createExternalLogMessage<TBody>(options: CreateExternalLogMessageOptions<TBody>): OutMessage
    builtin

/// \`UnsafeBodyNoRef\` is used to prevent default behavior: when message body is potentially large,
/// it's packed into a separate ref. Wrapping body with this struct tells the compiler:
/// "inline message body in-place, I guarantee it will fit".
/// \`\`\`
/// struct ProbablyLarge { a: (coins, coins, coins, coins, coins) }
///
/// val contents: ProbablyLarge = { ... };  // you are sure: values are small
/// createMessage({
///     // body: contents,           // by default, will be a ref (max 620 bits)
///     body: UnsafeBodyNoRef {
///         forceInline: contents,   // but it forces the compiler to inline it
///     }
/// \`\`\`
/// Another example: your body contains \`builder\` or \`RemainingBitsAndRefs\`: unpredictable size.
/// If you guarantee it's small and refs won't clash with code/data, avoid creating a cell.
struct UnsafeBodyNoRef<T> {
    forceInline: T
}

/// OutMessage is a result of [createMessage].
/// Essentially, it's a composed message cell, ready to be sent.
struct OutMessage {
    messageCell: cell
}

/// Sends a ready message cell.
/// For \`sendMode\`, see the constants above (SEND_MODE_*).
fun OutMessage.send(self, sendMode: int): void
    asm "SENDRAWMSG"

fun OutMessage.sendAndEstimateFee(self, sendMode: int): coins
    asm "SENDMSG"

@pure
fun OutMessage.estimateFeeWithoutSending(self, sendMode: int): coins
    asm "10 PUSHPOW2" "OR" "SENDMSG"

@pure
fun OutMessage.hash(self): uint256
    asm "HASHCU"


/// \`StateInit\` is a "canonical TL/B representation from block.tlb" of a contract initial state.
/// But for everyday tasks, it's too complicated. It is not used in practice.
/// To represent code+data, consider [ContractState].
/// To represent sharding (fixedPrefixLength, formerly splitDepth), consider [AutoDeployAddress].
struct StateInit {
    fixedPrefixLength: uint5?
    special: (bool, bool)?
    code: cell?
    data: cell?
    library: cell?
}

/// Calculates a hash of StateInit if only code+data are set.
/// \`\`\`
/// val addrHash = StateInit.calcHashCodeData(codeCell, dataCell);
/// createMessage({
///     dest: (workchain, addrHash),
///     ...
/// \`\`\`
@pure
fun StateInit.calcHashCodeData(code: cell, data: cell): uint256 asm (data code)
    "NEWC"
    "b{00110} STSLICECONST"
    "STREF"
    "STREF"
    "HASHBU"

/// Calculates a hash of StateInit if fixedPrefixLength+code+data are set.
@pure
fun StateInit.calcHashPrefixCodeData(fixedPrefixLength: uint5, code: cell, data: cell): uint256 asm (data code fixedPrefixLength)
    "NEWC"
    "b{1} STSLICECONST"
    "5 STU"
    "b{0110} STSLICECONST"
    "STREF"
    "STREF"
    "HASHBU"


/// Deprecated, use [address.calculateSameAddressInAnotherShard] instead.
/// Since late 2025 (TVM 12 has a cheap "builder-to-slice" instruction),
/// there is no need to use this function.
/// Historically its purpose was to avoid creating a cell and to return \`builder\` with a valid address.
@pure
@deprecated("use calculateSameAddressInAnotherShard instead")
fun address.buildSameAddressInAnotherShard(self, options: AddressShardingOptions): builder
    builtin

/// Given an internal address A="aaaa...a" returns "bbaa...a" (D bits from address B, 256-D from A).
///
/// Example for fixedPrefixLength (shard depth) = 8:
/// \`\`\`
/// self (A)     | aaaaaaaaaaa...aaa |
/// closeTo (B)  | 01010101bbb...bbb |   shardPrefix = 01010101 (depth 8)
/// result       | 01010101aaa...aaa |   address of A in same shard as B
/// \`\`\`
/// More precisely, self (input) is 267 bits: '100' (std addr no anycast) + workchainA + "aaaa...a".
/// The result is also 267 bits: '100' + workchainB + "bb" (D bits) + "aa...a" (256-D bits).
@pure
fun address.calculateSameAddressInAnotherShard(self, options: AddressShardingOptions): address
    builtin

/// Converts a builder containing a valid address to \`address\` (slice under the hood).
/// Deprecated! Since late 2025 (TVM 12 has a cheap "builder-to-slice" instruction),
/// all compiler built-ins return \`address\` directly, not \`builder\`.
@pure
@deprecated("prefer using \`address\` everywhere, avoid using \`builder\` and manual address construction")
fun address.fromValidBuilder(b: builder): address
    asm "ENDC" "CTOS"

/// Constructs an internal address (267 bits) from workchain and hash.
/// Hint: if you just want to send a message, don't call this function, pass workchain+hash directly:
/// \`\`\`
/// createMessage({
///     dest: (workchain, hash),   // a pair can be passed instead of \`address\`
///     ...                        // so, don't call this function in such a case
/// \`\`\`
fun address.fromWorkchainAndHash(workchain: int8, hash: uint256): address 
    asm (hash workchain)
    "NEWC"
    "b{100} STSLICECONST"       // addr_std$10 + 0 (no anycast)
    "8 STI"
    "256 STU"
    "BTOS"


/// Sends a raw message — a correctly serialized TL object \`Message X\`.
/// In practice, you'll use a high-level wrapper [createMessage].
fun sendRawMessage(msg: cell, mode: int): void
    asm "SENDRAWMSG"



/**
    Receiving and handling messages.
*/

/// \`InMessage\` is an input for \`onInternalMessage\` — when your contract accepts a non-bounced message.
/// Internally, some data exist on the stack, some can be acquired with TVM instructions.
/// The compiler replaces \`in.someField\` and gets a value correctly.
/// \`\`\`
/// fun onInternalMessage(in: InMessage) {
///     in.senderAddress   // actually calls \`INMSG_SRC\` asm instruction
///     in.body            // actually gets from a TVM stack
/// \`\`\`
struct InMessage {
    senderAddress: address          // an internal address from which the message arrived
    valueCoins: coins               // ton amount attached to an incoming message
    valueExtra: ExtraCurrenciesMap  // extra currencies attached to an incoming message
    originalForwardFee: coins       // fee that was paid by the sender
    createdLt: uint64               // logical time when a message was created
    createdAt: uint32               // unixtime when a message was created
    body: slice                     // message body, parse it with \`lazy AllowedMsg.fromSlice(in.body)\`
}

/// \`InMessageBounced\` is an input for \`onBouncedMessage\`.
/// Very similar to a non-bounced input [InMessage].
/// Note that \`bouncedBody\` is populated depending on [BounceMode] specified for [createMessage].
/// \`\`\`
/// fun onBouncedMessage(in: InMessageBounced) {
///     // in.bouncedBody is populated depending on BounceMode
///     // for instance, "Only256BitsOfBody" means 
///     // "bounced prefix 0xFFFFFFFF" and 256 bits of data
///     in.bouncedBody.skipBouncedPrefix();
///     val originalOpcode = in.bouncedBody.loadUint(32);
/// \`\`\`
struct InMessageBounced {
    senderAddress: address          // an internal address from which the message was bounced
    valueCoins: coins               // ton amount attached to a message
    valueExtra: ExtraCurrenciesMap  // extra currencies attached to a message
    originalForwardFee: coins       // comission that the sender has payed to send this message
    createdLt: uint64               // logical time when a message was created (and bounced)
    createdAt: uint32               // unixtime when a message was created (and bounced)
    bouncedBody: slice              // populated depending on \`BounceMode\` for \`createMessage\`
}

/// When you use [createMessage] with "rich bounce mode", and a message is bounced,
/// then \`in.bouncedBody\` inside \`onBouncedMessage\` has this format.
/// \`\`\`
/// fun onBouncedMessage(in: InMessageBounced) {
///     // suppose you ALWAYS use BounceMode.RichBounce in createMessage
///     // so you parse ALL bounces as follows
///     val rich = lazy RichBounceBody.fromSlice(in.bouncedBody);
///     // handle rich.originalBody, probably with lazy match
///     // use rich.xxx to get exitCode, gasUsed, and so on
/// \`\`\`
struct (0xfffffffe) RichBounceBody {
    originalBody: cell                              // a cell with original body, but for \`RichBounceOnlyRootCell\` without refs (cheaper)
    originalInfo: Cell<RichBounceOriginalMsgInfo>   // some fields of an original (outgoing) message
    bouncedByPhase: uint8                           // see docs
    exitCode: int32                                 // exception code from \`throw\` or TVM internals
    computePhase: RichBounceComputePhaseInfo?       // filled if compute phase executed
}

/// Describes compute phase details of a bounced message. 
struct RichBounceComputePhaseInfo {
    gasUsed: uint32
    vmSteps: uint32
}

/// Describes original outgoing message fields (besides body) of a bounced message.
/// \`\`\`
/// createMessage({
///     bounce: BounceMode.RichBounce,
///     value: ton("0.05")   // it will be \`valueCoins\` if this message is bounced
/// \`\`\`
struct RichBounceOriginalMsgInfo {
    valueCoins: coins
    valueExtra: ExtraCurrenciesMap
    createdLt: uint64
    createdAt: uint32
}

/// Skip 0xFFFFFFFF prefix (when a message is bounced).
/// Used when parsing \`in.bounceBody\` for "not rich" \`BounceMode.Only256BitsOfBody\` bounces.
@pure
fun slice.skipBouncedPrefix(mutate self): self
    asm "32 LDU" "NIP"

/// Load msgFlags from incoming message body (4 bits).
@pure
@deprecated("use \`onBouncedMessage\` handler instead of parsing msgCell flags manually")
fun slice.loadMessageFlags(mutate self): int
    asm( -> 1 0) "4 LDU"

/// Loads a message opcode (32-bit integer) when parsing message body manually.
@pure
fun slice.loadMessageOp(mutate self): int
    asm( -> 1 0) "32 LDU"

/// Stores a message opcode (32-bit integer) when composing an output message manually.
@pure
@deprecated("use \`createMessage\` instead of composing messages manually")
fun builder.storeMessageOp(mutate self, op: int): self
    asm(op self) "32 STU"

/// Loads a message queryId (64-bit integer, immediately following the opcode) when parsing message body manually.
@pure
@deprecated("use structures and lazy match instead of parsing messages manually")
fun slice.loadMessageQueryId(mutate self): int
    asm( -> 1 0) "64 LDU"

@pure
@deprecated("use structures and lazy loading instead of parsing messages manually")
fun slice.skipMessageQueryId(mutate self): self
    asm "64 LDU" "NIP"

@pure
@deprecated("use \`createMessage\` instead of composing messages manually")
fun builder.storeMessageQueryId(mutate self, queryId: int): self
    asm(queryId self) "64 STU"


`,

'@stdlib/exotic-cells.tolk':`// A part of standard library for Tolk
tolk 1.3

enum ExoticCellType: int8 {
    Ordinary = -1
    PrunedBranch = 1
    LibraryReference = 2
    MerkleProof = 3
    MerkleUpdate = 4
}

/// Transforms an ordinary or exotic cell into a slice, as if it were an ordinary cell. 
/// A flag is returned indicating whether c is exotic. 
/// If so, ExoticCellType can later be deserialized from the first eight bits of s.
/// \`\`\`
/// var (s, isExotic) = c.beginParseSpecial();
/// if (isExotic) {
///     // use unsafe cast because the first 8 bits hold a correct value, no need for validation
///     val cellType = s.loadInt(8) as ExoticCellType;
/// }
/// \`\`\`
@pure
fun cell.beginParseSpecial(self): (slice, bool)
    asm "XCTOS"

/// Creates an exotic or ordinary cell from a builder.
@pure
fun builder.endCellSpecial(self, isExotic: bool = true): cell
    asm "ENDXC"

/// Returns i-th hash of the cell.
@pure
fun cell.hashOfLevel(self, level: int): uint256
    asm "CHASHIX"

/// Returns i-th depth of the cell.
@pure
fun cell.depthOfLevel(self, level: int): int
    asm "CDEPTHIX"

/// Loads an exotic cell and returns an ordinary cell.
/// If the cell is already ordinary, does nothing. Throws an exception if cannot be loaded.
@pure
fun cell.loadSpecial(self): cell
    asm "XLOAD"

/// Transforms an ordinary or exotic cell into a library reference
/// (into an exotic cell with contents "8 bits cell type" + "256 bits lib hash").
@pure
fun cell.toLibraryReference(self): cell {
    return beginCell()
        .storeInt(ExoticCellType.LibraryReference as int, 8)
        .storeUint(self.hash(), 256)
        .endCellSpecial()
}
`,

'@stdlib/gas-payments.tolk':`// A part of standard library for Tolk
tolk 1.3

/**
  Gas and payment related primitives.
 */

/// Returns amount of gas (in gas units) consumed in current Computation Phase.
fun getGasConsumedAtTheMoment(): int
    asm "GASCONSUMED"

/// This function is required to be called when you process an external message (from an outer world)
/// and "accept" it to blockchain.
/// Without calling this function, an external message would be discarded.
/// As an effect, the current smart contract agrees to buy some gas to finish the current transaction.
/// For more details, check [accept_message effects](https://ton.org/docs/#/smart-contracts/accept).
fun acceptExternalMessage(): void
    asm "ACCEPT"

/// When processing an internal message, by default, the limit of gas consumption is determined by incoming message.
/// Functions [setGasLimit] and [setGasLimitToMaximum] allow you to change this behavior.
/// Sets current gas limit \`gl\` to its maximal allowed value \`gm\`, and resets the gas credit \`gc\` to zero,
/// decreasing the value of \`gr\` by \`gc\` in the process.
fun setGasLimitToMaximum(): void
    asm "ACCEPT"

/// When processing an internal message, by default, the limit of gas consumption is determined by incoming message.
/// Functions [setGasLimit] and [setGasLimitToMaximum] allow you to change this behavior.
/// Sets current gas limit \`gl\` to the minimum of limit and \`gm\`, and resets the gas credit \`gc\` to zero.
/// If the gas consumed so far (including the present instruction) exceeds the resulting value of \`gl\`,
/// an (unhandled) out of gas exception is thrown before setting new gas limits.
fun setGasLimit(limit: int): void
    asm "SETGASLIMIT"

/// Calculates fee (amount in nanotoncoins to be paid) for a transaction which consumed \`gasUsed\` gas units.
@pure
fun calculateGasFee(workchain: int8, gasUsed: int): coins
    asm(gasUsed workchain) "GETGASFEE"

/// Same as [calculateGasFee], but without flat price. [Docs about fees](https://docs.ton.org/develop/howto/fees-low-level)
@pure
fun calculateGasFeeWithoutFlatPrice(workchain: int8, gasUsed: coins): coins
    asm(gasUsed workchain) "GETGASFEESIMPLE"

/// Calculates amount of nanotoncoins you should pay for storing a contract of provided size for \`seconds\`.
/// \`bits\` and \`cells\` represent contract state (code + data).
fun calculateStorageFee(workchain: int8, seconds: int, bits: int, cells: int): coins
    asm(cells bits seconds workchain) "GETSTORAGEFEE"

/// Calculates amount of nanotoncoins you should pay to send a message of a specified size.
@pure
fun calculateForwardFee(workchain: int8, bits: int, cells: int): coins
    asm(cells bits workchain) "GETFORWARDFEE"

/// Same as [calculateForwardFee], but without lump price. [Docs about fees](https://docs.ton.org/develop/howto/fees-low-level)
@pure
fun calculateForwardFeeWithoutLumpPrice(workchain: int8, bits: int, cells: int): coins
    asm(cells bits workchain) "GETFORWARDFEESIMPLE"

/// Calculates fee that was paid by the sender of an incoming internal message.
@deprecated("use modern \`onInternalMessage\` and access \`in.originalForwardFee\` directly")
fun calculateOriginalForwardFee(workchain: int8, incomingFwdFee: coins): coins
    asm(incomingFwdFee workchain) "GETORIGINALFWDFEE"

/// Returns the amount of nanotoncoins current contract debts for storage. ("due" and "debt" are synonyms)
/// If it has no debt, \`0\` is returned.
fun contract.getStorageDuePayment(): coins
    asm "DUEPAYMENT"

/// Returns the amount of nanotoncoins charged for storage.
/// (during storage phase preceeding to current computation phase)
@pure
fun contract.getStoragePaidPayment(): coins
    asm "STORAGEFEES"
`,

'@stdlib/lisp-lists.tolk':`// A part of standard library for Tolk
tolk 1.3

/**
    A lisp-style list is a set of nested 2-element TVM tuples:
    \`[1, [2, [3, null]]]\` represents a list \`[1, 2, 3]\`.
    Unlike \`array<T>\`, it can represent a list longer than 255 elements,
    but it has access only to the frontmost element (head).

    To create a lisp list, use \`[]\` with a hint:
    \`\`\`
    var emptyList: lisp_list<int> = [];
    // or
    var emptyList = [] as lisp_list<int>;
    \`\`\`

    Creating non-empty lists is also possible:
    \`\`\`
    var numsList = [1, 2, 3] as lisp_list<int>;
    // stack: [1 [2 [3 null]]]
    \`\`\`
 */

@pure
fun lisp_list<T>.isEmpty(self): bool {
    return self.tvmTuple == null
}

/// Inserts an element an the front.
/// Before: \`[a b c]\`.
/// After: \`[new a b c]\`.
@pure
fun lisp_list<T>.prependHead(mutate self, value: T): void {
    self.tvmTuple = [value, self.tvmTuple as unknown as lisp_list<T>];
}

/// Pops an element from the front and returns it.
/// Before: \`[a b c]\`.
/// After: \`[b c]\`, returns \`a\`.
/// Throws an exceptions if a list is empty.
@pure
fun lisp_list<T>.popHead(mutate self): T {
    var [head, tail] = self.tvmTuple!;
    self.tvmTuple = tail as unknown as [T, lisp_list<T>]?;
    return head
}

/// Returns an element from the front: \`a\` for \`[a b c]\`.
/// Throws an exceptions if a list is empty.
@pure
fun lisp_list<T>.getHead(self): T {
    return self.tvmTuple!.0
}

/// Returns all but the frontmost element: \`[b c]\` for \`[a b c]\`.
/// Throws an exceptions if a list is empty.
@pure
fun lisp_list<T>.getTail(self): lisp_list<T> {
    return self.tvmTuple!.1
}

/// Calculates the size of a list by looping all the elements (there is no other way to do it).
/// The longer the list — the more gas consumption (O(N) complexity).
@pure
fun lisp_list<T>.calculateSize(self): int {
    var size = 0;
    var ptr = self;
    while (!ptr.isEmpty()) {
        ptr = ptr.getTail();
        size += 1;
    }
    return size;
}

/// Returns the list with elements in reversed order.
/// Given \`[a b c]\` returns \`[c b a]\`.
@pure
fun lisp_list<T>.calculateReversed(self): lisp_list<T> {
    var reversed = [] as lisp_list<T>;
    var ptr = self;
    while (!ptr.isEmpty()) {
        reversed.prependHead(ptr.popHead());
    }
    return reversed;
}

/// Returns the concatenation of two lists.
/// Given \`[a b]\` and \`[c d]\`, returns \`[a b c d]\`.
@pure
fun lisp_list<T>.calculateConcatenation(self, other: lisp_list<T>): lisp_list<T> {
    var reversed = [] as lisp_list<T>;
    var ptr = self;
    while (!ptr.isEmpty()) {
        reversed.prependHead(ptr.popHead());
    }
    ptr = other;
    while (!ptr.isEmpty()) {
        reversed.prependHead(ptr.popHead());
    }
    return reversed.calculateReversed();
}

/// Creates a lisp_list<T> with zero elements.
/// Deprecated. Use \`[]\` syntax instead:
/// \`\`\`
/// // old way, deprecated:
/// var list = createEmptyList<int>();
///
/// // new way, preferred:
/// var t = lisp_list<int> [];
/// \`\`\`
@pure
@deprecated("use \`[]\` instead of \`createEmptyList\`")
fun createEmptyList<T>(): lisp_list<T>
    asm "PUSHNULL"


/// \`lisp_list<T>\` can be serialized if \`T\` is serializable. It's encoded as reversed snake refs:
/// A list \`[1 2 3]\` (on a stack: \`[1 [2 [3 null]]]\`) is "3.ref(2.ref(1.ref(empty cell)))".
/// An empty list \`[]\` (on a stack: \`null\`) is "empty cell"
fun lisp_list<T>.packToBuilder(self, mutate b: builder): void {
    var tail = createEmptyCell();
    var ptr = self;
    while (!ptr.isEmpty()) {
        tail = beginCell().storeRef(tail).storeAny<T>(ptr.popHead()).endCell();
    }
    b.storeRef(tail);
}

/// Unserialize \`lisp_list<T>\` from snake refs:
/// From "3.ref(2.ref(1.ref(empty cell)))" get a list \`[1 2 3]\` (on a stack: \`[1 [2 [3 null]]]\`).
fun lisp_list<T>.unpackFromSlice(mutate s: slice): lisp_list<T> {
    var result = [] as lisp_list<T>;
    var head = s.loadRef().beginParse();
    while (!head.isEmpty()) {
        val tailSnaked = head.loadRef();
        val headValue = head.loadAny<T>();
        head.assertEnd();   // ensure no data is present besides T
        result.prependHead(headValue);
        head = tailSnaked.beginParse();
    }
    return result;
}
`,

'@stdlib/reflection.tolk':`// A part of standard library for Tolk
tolk 1.3

/**
    Reflection API.
    \`reflect\` is a built-in struct with only static methods for compile-time type introspection.
 */

/// \`reflect\` is a built-in struct, it has only static methods.
/// Example: \`reflect.typeNameOf<int>()\` and other methods.
struct reflect

/// Get human-readable type name of any T for logging or low-level purposes.
/// Example:
/// \`\`\`
/// debug.printString(reflect.typeNameOf<int>());  // "int"
/// \`\`\`
@pure
fun reflect.typeNameOf<T>(): string
    builtin

/// Get human-readable type name of any object for logging or low-level purposes.
/// Example:
/// \`\`\`
/// debug.printString(reflect.typeNameOfObject(42));   // "int"
/// \`\`\`
@pure
fun reflect.typeNameOfObject<T>(anyVariable: T): string
    builtin

/// Returns the number of stack slots a type T occupies.
/// Example:
/// \`\`\`
/// reflect.stackSizeOf<int>()     // 1
/// reflect.stackSizeOf<(int, int)>()  // 2
/// \`\`\`
@pure
fun reflect.stackSizeOf<T>(): int
    builtin

/// Returns the number of stack slots any object occupies.
/// - for nullableInt it's 1, because \`int?\` is 1 TVM slot holding either NULL or a value.
/// - for somePoint it's 2 for \`struct Point { x:int, y:int }\`: two fields one slot per each.
/// Useful for debugging or when preparing stack contents for RUNVM.
@pure
fun reflect.stackSizeOfObject<T>(anyVariable: T): int
    builtin

/// Returns serialization prefix of a struct.
/// Example: for \`struct (0xF0) AssetRegular { ... }\` will return \`(240, 8)\`.
/// \`\`\`
/// reflect.serializationPrefixOf<AssetRegular>().1   // 8
/// \`\`\`
@pure
fun reflect.serializationPrefixOf<T>(): (int, int)
    builtin

/// Estimate serialization size of any type.
/// Returns (minBits, maxBits, minRefs, maxRefs).
/// Example:
/// \`\`\`
/// reflect.estimateSerializationOf<int8?>()     // (1 9 0 0)
/// \`\`\`
@pure
fun reflect.estimateSerializationOf<T>(): (int, int, int, int)
    builtin

/// Source location information, resolved at compile time.
/// See [reflect.sourceLocation] for examples.
struct SourceLocation {
    lineNo: int
    charNo: int
    fileName: string
}

/// Returns the source location where this function is called.
/// Can be used to retrieve current file/line, similar to __FILE__ in some languages.
/// Example:
/// \`\`\`
/// var curLineNo = reflect.sourceLocation().lineNo
/// \`\`\`
/// When used as a default parameter value, captures the caller's location.
/// Example:
/// \`\`\`
/// fun log(msg: string, loc: SourceLocation = reflect.sourceLocation()) {
///     debug.print(loc.lineNo);
/// }
///
/// fun demo() {
///     log("a");    // prints K — current line no
///     log("b");    // prints K+1
/// }
/// \`\`\`
@pure
fun reflect.sourceLocation(): SourceLocation
    builtin

/// Behaves like [reflect.sourceLocation], but returns a string "fileName:lineNo:charNo".
/// \`fileName\` contains the full (real) path of the source file.
/// Being used as a default parameter value, captures the caller's location.
@pure
fun reflect.sourceLocationAsString(): string
    builtin
`,

'@stdlib/strings.tolk':`// A part of standard library for Tolk
tolk 1.3

/**
    A string is a language primitive — a cell containing snake-formatted text data.
    Short strings are (typically) represented as a cell, like "abcd".
    Since the limit of a cell is 1023 bits (127 bytes), the only way to represent longer strings
  is to use snake format: "start" + (ref "middle" + (ref "end")).
    Short strings can also be stored in snake format, e.g. "ab" + (ref "cd").
    The outer world, knowing that a particular cell is a string, can iteratively concatenate
  all references and render a string in UI.
 */

/// Returns snake depth of a string.
/// Example: "abcd" has depth 0, it's a flat string.
/// Example: "ab" + (ref "cd") has depth 1.
/// Example: "a" + (ref "bc" + (ref "d" + (ref ""))) has depth 3.
@pure
fun string.depth(self): int
    asm "CDEPTH"

/// Calculates length of a string in bytes (not unicode symbols, just bytes).
/// Since strings are stored in snake format (chain of cells), this requires traversing all cells.
@pure
@inline_ref
fun string.calculateLength(self): int {
    var bits = 0;
    var snake = self.beginParse();
    repeat (snake.depth()) {
        bits += snake.remainingBitsCount();
        snake = snake.loadRef().beginParse();
    }
    bits += snake.remainingBitsCount();
    return bits / 8;
}

/// Computes a content hash of a string, traversing all snake cells.
/// The hash is the same regardless of snake layout:
/// "abcd", "ab" + (ref "cd"), "a" + (ref "bcd" + (ref "")) all produce the same hash.
/// Uses FNV-1a hashing with standard 128-bit parameters.
@pure
@inline_ref
fun string.hash(self): uint256 {
    val FNV_PRIME = 309485009821345068724781371;
    val MASK = (1 << 128) - 1;
    var snake = self.beginParse();
    var hash = 144066263297769815596495629667062367629;
    repeat (snake.depth()) {
        repeat (snake.remainingBitsCount() / 8) {
            hash = ((hash ^ snake.loadUint(8)) * FNV_PRIME) & MASK;
        }
        snake = snake.loadRef().beginParse();
    }
    repeat (snake.remainingBitsCount() / 8) {
        hash = ((hash ^ snake.loadUint(8)) * FNV_PRIME) & MASK;
    }
    return hash;
}

/// Compares two strings, considering snake cells.
/// For example, "abcd" is equal to "ab" + (ref "cd") is equal to "a" + (ref "bc" + (ref "d")).
@pure
fun string.equalTo(self, another: string): bool {
    return self.depth() + another.depth()
        // some is snake, so we have "ab->cd" / "a->bc->d" / etc., do full traverse
        ? self.hash() == another.hash()
        // fast path for flat strings (no refs), use native cell hashes
        : (self as unknown as cell).hash() == (another as unknown as cell).hash()
}

/// Converts an integer to its decimal string representation.
/// Returns a string (a TVM cell on a stack).
fun int.toDecimalString(self): string asm """
    NEWC
    SWAP                      // b x
    DUP                       // b x x
    ISNEG
    IF:<{                     // if x<0
      SWAP                    // x b
      x{2D} STSLICECONST      // store '-' = 45 = 0x2D
      SWAP                    // b x
      NEGATE                  // b |x|
    }>
    // now: b x (x is always >= 0)
    0 PUSHINT                 // b x i=0
    SWAP                      // b i=0 x
    UNTIL:<{                  // loop: extract digits from x, push them under builder
      10 PUSHINT DIVMOD       // b i x/10 x%10
      48 ADDCONST             // b i x digit (digit = '0' + remainder)
      s3 s1 s3 XCHG3          // digit b x i (digit goes under builder)
      INC                     // digit b x i+1
      s1 s0 XCPU              // digit b i x x (copy x to check if zero)
      ISZERO                  // exit when x == 0
    }>
    DROP                      // remove x (now 0)
    // stack: d1 d2 ... dN b i (digits in correct order, oldest at bottom)
    REPEAT:<{ 8 STU }>        // store i digits (8 bits each) into builder
    ENDC                      // finalize cell
"""

/// A utility for building strings at runtime by appending multiple parts.
/// The result is a valid snake-format string.
/// \`\`\`
/// var s = StringBuilder.create().append("Hello, ").append(userName).append("!").build();
/// \`\`\`
/// After this, \`s\` will be a chain of cells: "Hello, " + (ref userName + (ref "!"))
struct StringBuilder {
    private chunks: array<string> = []
}

/// Creates an empty \`StringBuilder\`.
/// \`\`\`
/// var s = StringBuilder.create().append(...).build();
/// \`\`\`
fun StringBuilder.create(): StringBuilder {
    return {}
}

/// Appends a string to the builder. The string may itself be a snake chain.
fun StringBuilder.append(mutate self, s: string): self {
    self.chunks.push(s);
    return self;
}

/// Appends an integer as its decimal representation.
/// For example, \`sb.appendInt(-42)\` appends "-42".
@inline_ref
fun StringBuilder.appendInt(mutate self, n: int): self {
    self.chunks.push(n.toDecimalString());
    return self;
}

/// Builds the final string by concatenating all appended parts into a snake chain.
/// The result is a valid snake-format string where each cell contains data and optionally a ref to next.
/// Throws an exception if \`append()\` was not called (a builder is empty).
@inline_ref
fun StringBuilder.build(self): string {
    // the last string, as any other, is probably "xyz", probably snaked "x->y->z"
    // we'll leave it as-is (TVM cell), won't even call \`beginParse\` for it
    var tail = self.chunks.last() as unknown as cell;
    var idx = self.chunks.size() - 1;
    // we can have simple chunks [ "ab", "cd", "xy" ], it's the hot path, loop backwards
    // but some middle strings might be snaked: [ "ab->c->d", "xy->z" ], we have to unwrap them
    repeat (idx) {
        idx -= 1;
        var chunk = self.chunks.get(idx).beginParse();
        // in case the middle string is a snake: "ab->c->d" + tail
        if (chunk.depth()) {
            // split "ab->c->d" into [ "ab", "c", "d" ]
            var snake = chunk;
            var subBits = array<slice> [];
            repeat (snake.depth()) {
                var nextSnake = snake.loadRef().beginParse();
                subBits.push(snake);
                snake = nextSnake;
            }
            subBits.push(snake);
            // calculate { tail = "d" + tail; tail = "c" + tail }
            var lastIdx = subBits.size() - 1;
            repeat (lastIdx) {
                tail = (subBits.get(lastIdx), tail).toCell();
                lastIdx -= 1;
            }
            // save chunk = "ab", below it will be concatenated "ab" + tail
            chunk = subBits.first();
        }
        // tail = "cur_str" + tail
        tail = (chunk, tail).toCell();
    }
    // return a TVM cell, which is a valid snake string
    return tail as unknown as string;
}

/// A string, which contents starts with 0x00/0x01 byte.
/// By standard, offchain explorers/APIs may treat contents of a string based on 0xXX prefixes:
/// - 0x00 + "str" — means "this is a regular snake-encoded string"
/// - 0x01 + "str" — means "this is an off-chain content URL"
/// Used in jettons and NFTs for responses from GET methods to comply with TEPs (token standards).
type string_prefixed0x = string

/// Given "str", prefix it with 0x00 byte.
/// Logically it makes a string "\\x00str"; physically, it's 0x00 + ref "str".
/// See [string_prefixed0x].
fun string.prefixWith00(self): string_prefixed0x {
    val cell00 = beginCell().storeUint(0x00, 8).storeString(self).endCell();
    return cell00 as unknown as string;
}

/// Given "str", prefix it with 0x01 byte.
/// Logically it makes a string "\\x01str"; physically, it's 0x01 + ref "str".
/// See [string_prefixed0x].
fun string.prefixWith01(self): string_prefixed0x {
    val cell01 = beginCell().storeUint(0x01, 8).storeString(self).endCell();
    return cell01 as unknown as string;
}
`,

'@stdlib/tvm-dicts.tolk':`// A part of standard library for Tolk
tolk 1.3

/**
    Low-level API working with TVM dictionaries.
    Not recommended to use, since it's very complicated and error-prone.
    Use \`map<K, V>\` instead, it's even more efficient then low-level dict API.

    Dictionaries are represented as \`cell\` data type.
    Most of functions are duplicated for three common cases:
    - iDict* - dicts with signed integer keys
    - uDict* - dicts with unsigned integer keys
    - sDict* - dicts with arbitrary slice keys
    When accessing a dict element, you should not only provide a key, but provide keyLen,
  since for optimization, key length is not stored in the dictionary itself.
    Every dictionary object (\`self\` parameter) can be null. TVM NULL is essentially "empty dictionary".
 */

/// In @stdlib/common.tolk, there is a type alias:
/// \`\`\`
/// type dict = cell?
/// \`\`\`
/// For clarity, we use "dict" instead of a "cell?" where a cell-dictionary is assumed.

/// Creates an empty dictionary, which is actually a null value. Equivalent to \`PUSHNULL\`
@pure
fun createEmptyDict(): dict
    asm "NEWDICT"

/// Checks whether a dictionary is empty.
@pure
fun dict.dictIsEmpty(self): bool
    asm "DICTEMPTY"


@pure
fun dict.iDictGet(self, keyLen: int, key: int): (slice?, bool)
    asm(key self keyLen) "DICTIGET" "NULLSWAPIFNOT"

@pure
fun dict.uDictGet(self, keyLen: int, key: int): (slice?, bool)
    asm(key self keyLen) "DICTUGET" "NULLSWAPIFNOT"

@pure
fun dict.sDictGet(self, keyLen: int, key: slice): (slice?, bool)
    asm(key self keyLen) "DICTGET" "NULLSWAPIFNOT"


@pure
fun dict.iDictSet(mutate self, keyLen: int, key: int, value: slice): void
    asm(value key self keyLen) "DICTISET"

@pure
fun dict.uDictSet(mutate self, keyLen: int, key: int, value: slice): void
    asm(value key self keyLen) "DICTUSET"

@pure
fun dict.sDictSet(mutate self, keyLen: int, key: slice, value: slice): void
    asm(value key self keyLen) "DICTSET"


@pure
fun dict.iDictSetRef(mutate self, keyLen: int, key: int, value: cell): void
    asm(value key self keyLen) "DICTISETREF"

@pure
fun dict.uDictSetRef(mutate self, keyLen: int, key: int, value: cell): void
    asm(value key self keyLen) "DICTUSETREF"

@pure
fun dict.sDictSetRef(mutate self, keyLen: int, key: slice, value: cell): void
    asm(value key self keyLen) "DICTSETREF"


@pure
fun dict.iDictSetIfNotExists(mutate self, keyLen: int, key: int, value: slice): bool
    asm(value key self keyLen) "DICTIADD"

@pure
fun dict.uDictSetIfNotExists(mutate self, keyLen: int, key: int, value: slice): bool
    asm(value key self keyLen) "DICTUADD"


@pure
fun dict.iDictSetIfExists(mutate self, keyLen: int, key: int, value: slice): bool
    asm(value key self keyLen) "DICTIREPLACE"

@pure
fun dict.uDictSetIfExists(mutate self, keyLen: int, key: int, value: slice): bool
    asm(value key self keyLen) "DICTUREPLACE"


@pure
fun dict.iDictGetRef(self, keyLen: int, key: int): (cell?, bool)
    asm(key self keyLen) "DICTIGETREF" "NULLSWAPIFNOT"

@pure
fun dict.uDictGetRef(self, keyLen: int, key: int): (cell?, bool)
    asm(key self keyLen) "DICTUGETREF" "NULLSWAPIFNOT"

@pure
fun dict.sDictGetRef(self, keyLen: int, key: slice): (cell?, bool)
    asm(key self keyLen) "DICTGETREF" "NULLSWAPIFNOT"


@pure
fun dict.iDictGetRefOrNull(self, keyLen: int, key: int): cell?
    asm(key self keyLen) "DICTIGETOPTREF"

@pure
fun dict.uDictGetRefOrNull(self, keyLen: int, key: int): cell?
    asm(key self keyLen) "DICTUGETOPTREF"

@pure
fun dict.sDictGetRefOrNull(self, keyLen: int, key: slice): cell?
    asm(key self keyLen) "DICTGETOPTREF"


@pure
fun dict.iDictDelete(mutate self, keyLen: int, key: int): bool
    asm(key self keyLen) "DICTIDEL"

@pure
fun dict.uDictDelete(mutate self, keyLen: int, key: int): bool
    asm(key self keyLen) "DICTUDEL"

@pure
fun dict.sDictDelete(mutate self, keyLen: int, key: slice): bool
    asm(key self keyLen) "DICTDEL"


@pure
fun dict.iDictSetAndGet(mutate self, keyLen: int, key: int, value: slice): (slice?, bool)
    asm(value key self keyLen) "DICTISETGET" "NULLSWAPIFNOT"

@pure
fun dict.uDictSetAndGet(mutate self, keyLen: int, key: int, value: slice): (slice?, bool)
    asm(value key self keyLen) "DICTUSETGET" "NULLSWAPIFNOT"

@pure
fun dict.sDictSetAndGet(mutate self, keyLen: int, key: slice, value: slice): (slice?, bool)
    asm(value key self keyLen) "DICTSETGET" "NULLSWAPIFNOT"


@pure
fun dict.iDictSetAndGetRefOrNull(mutate self, keyLen: int, key: int, value: cell): cell?
    asm(value key self keyLen) "DICTISETGETOPTREF"

@pure
fun dict.uDictSetAndGetRefOrNull(mutate self, keyLen: int, key: int, value: cell): cell?
    asm(value key self keyLen) "DICTUSETGETOPTREF"


@pure
fun dict.iDictDeleteAndGet(mutate self, keyLen: int, key: int): (slice?, bool)
    asm(key self keyLen) "DICTIDELGET" "NULLSWAPIFNOT"

@pure
fun dict.uDictDeleteAndGet(mutate self, keyLen: int, key: int): (slice?, bool)
    asm(key self keyLen) "DICTUDELGET" "NULLSWAPIFNOT"

@pure
fun dict.sDictDeleteAndGet(mutate self, keyLen: int, key: slice): (slice?, bool)
    asm(key self keyLen) "DICTDELGET" "NULLSWAPIFNOT"


@pure
fun dict.iDictSetBuilder(mutate self, keyLen: int, key: int, value: builder): void
    asm(value key self keyLen) "DICTISETB"

@pure
fun dict.uDictSetBuilder(mutate self, keyLen: int, key: int, value: builder): void
    asm(value key self keyLen) "DICTUSETB"

@pure
fun dict.sDictSetBuilder(mutate self, keyLen: int, key: slice, value: builder): void
    asm(value key self keyLen) "DICTSETB"


@pure
fun dict.iDictSetBuilderIfNotExists(mutate self, keyLen: int, key: int, value: builder): bool
    asm(value key self keyLen) "DICTIADDB"

@pure
fun dict.uDictSetBuilderIfNotExists(mutate self, keyLen: int, key: int, value: builder): bool
    asm(value key self keyLen) "DICTUADDB"

@pure
fun dict.iDictSetBuilderIfExists(mutate self, keyLen: int, key: int, value: builder): bool
    asm(value key self keyLen) "DICTIREPLACEB"

@pure
fun dict.uDictSetBuilderIfExists(mutate self, keyLen: int, key: int, value: builder): bool
    asm(value key self keyLen) "DICTUREPLACEB"


@pure
fun dict.iDictDeleteFirstAndGet(mutate self, keyLen: int): (int?, slice?, bool)
    asm(-> 0 2 1 3) "DICTIREMMIN" "NULLSWAPIFNOT2"

@pure
fun dict.uDictDeleteFirstAndGet(mutate self, keyLen: int): (int?, slice?, bool)
    asm(-> 0 2 1 3) "DICTUREMMIN" "NULLSWAPIFNOT2"

@pure
fun dict.sDictDeleteFirstAndGet(mutate self, keyLen: int): (slice?, slice?, bool)
    asm(-> 0 2 1 3) "DICTREMMIN" "NULLSWAPIFNOT2"


@pure
fun dict.iDictDeleteLastAndGet(mutate self, keyLen: int): (int?, slice?, bool)
    asm(-> 0 2 1 3) "DICTIREMMAX" "NULLSWAPIFNOT2"

@pure
fun dict.uDictDeleteLastAndGet(mutate self, keyLen: int): (int?, slice?, bool)
    asm(-> 0 2 1 3) "DICTUREMMAX" "NULLSWAPIFNOT2"

@pure
fun dict.sDictDeleteLastAndGet(mutate self, keyLen: int): (slice?, slice?, bool)
    asm(-> 0 2 1 3) "DICTREMMAX" "NULLSWAPIFNOT2"


@pure
fun dict.iDictGetFirst(self, keyLen: int): (int?, slice?, bool)
    asm (-> 1 0 2) "DICTIMIN" "NULLSWAPIFNOT2"

@pure
fun dict.uDictGetFirst(self, keyLen: int): (int?, slice?, bool)
    asm (-> 1 0 2) "DICTUMIN" "NULLSWAPIFNOT2"

@pure
fun dict.sDictGetFirst(self, keyLen: int): (slice?, slice?, bool)
    asm (-> 1 0 2) "DICTMIN" "NULLSWAPIFNOT2"

@pure
fun dict.iDictGetFirstAsRef(self, keyLen: int): (int?, cell?, bool)
    asm (-> 1 0 2) "DICTIMINREF" "NULLSWAPIFNOT2"

@pure
fun dict.uDictGetFirstAsRef(self, keyLen: int): (int?, cell?, bool)
    asm (-> 1 0 2) "DICTUMINREF" "NULLSWAPIFNOT2"

@pure
fun dict.sDictGetFirstAsRef(self, keyLen: int): (slice?, cell?, bool)
    asm (-> 1 0 2) "DICTMINREF" "NULLSWAPIFNOT2"


@pure
fun dict.iDictGetLast(self, keyLen: int): (int?, slice?, bool)
    asm (-> 1 0 2) "DICTIMAX" "NULLSWAPIFNOT2"

@pure
fun dict.uDictGetLast(self, keyLen: int): (int?, slice?, bool)
    asm (-> 1 0 2) "DICTUMAX" "NULLSWAPIFNOT2"

@pure
fun dict.sDictGetLast(self, keyLen: int): (slice?, slice?, bool)
    asm (-> 1 0 2) "DICTMAX" "NULLSWAPIFNOT2"

@pure
fun dict.iDictGetLastAsRef(self, keyLen: int): (int?, cell?, bool)
    asm (-> 1 0 2) "DICTIMAXREF" "NULLSWAPIFNOT2"

@pure
fun dict.uDictGetLastAsRef(self, keyLen: int): (int?, cell?, bool)
    asm (-> 1 0 2) "DICTUMAXREF" "NULLSWAPIFNOT2"

@pure
fun dict.sDictGetLastAsRef(self, keyLen: int): (slice?, cell?, bool)
    asm (-> 1 0 2) "DICTMAXREF" "NULLSWAPIFNOT2"


@pure
fun dict.iDictGetNext(self, keyLen: int, pivot: int): (int?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTIGETNEXT" "NULLSWAPIFNOT2"

@pure
fun dict.uDictGetNext(self, keyLen: int, pivot: int): (int?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTUGETNEXT" "NULLSWAPIFNOT2"

@pure
fun dict.sDictGetNext(self, keyLen: int, pivot: slice): (slice?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTGETNEXT" "NULLSWAPIFNOT2"

@pure
fun dict.iDictGetNextOrEqual(self, keyLen: int, pivot: int): (int?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTIGETNEXTEQ" "NULLSWAPIFNOT2"

@pure
fun dict.uDictGetNextOrEqual(self, keyLen: int, pivot: int): (int?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTUGETNEXTEQ" "NULLSWAPIFNOT2"

@pure
fun dict.sDictGetNextOrEqual(self, keyLen: int, pivot: slice): (slice?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTGETNEXTEQ" "NULLSWAPIFNOT2"


@pure
fun dict.iDictGetPrev(self, keyLen: int, pivot: int): (int?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTIGETPREV" "NULLSWAPIFNOT2"

@pure
fun dict.uDictGetPrev(self, keyLen: int, pivot: int): (int?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTUGETPREV" "NULLSWAPIFNOT2"

@pure
fun dict.sDictGetPrev(self, keyLen: int, pivot: slice): (slice?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTGETPREV" "NULLSWAPIFNOT2"

@pure
fun dict.iDictGetPrevOrEqual(self, keyLen: int, pivot: int): (int?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTIGETPREVEQ" "NULLSWAPIFNOT2"

@pure
fun dict.uDictGetPrevOrEqual(self, keyLen: int, pivot: int): (int?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTUGETPREVEQ" "NULLSWAPIFNOT2"

@pure
fun dict.sDictGetPrevOrEqual(self, keyLen: int, pivot: slice): (slice?, slice?, bool)
    asm(pivot self keyLen -> 1 0 2) "DICTGETPREVEQ" "NULLSWAPIFNOT2"


/**
    Prefix dictionary primitives.
 */

@pure
fun dict.prefixDictGet(self, keyLen: int, key: slice): (slice, slice?, slice?, bool)
    asm(key self keyLen) "PFXDICTGETQ" "NULLSWAPIFNOT2"

@pure
fun dict.prefixDictSet(mutate self, keyLen: int, key: slice, value: slice): bool
    asm(value key self keyLen) "PFXDICTSET"

@pure
fun dict.prefixDictDelete(mutate self, keyLen: int, key: slice): bool
    asm(key self keyLen) "PFXDICTDEL"
`,

'@stdlib/tvm-lowlevel.tolk':`// A part of standard library for Tolk
tolk 1.3

/// Usually \`c3\` has a continuation initialized by the whole code of the contract. It is used for function calls.
/// The primitive returns the current value of \`c3\`.
@pure
fun getTvmRegisterC3(): continuation
    asm "c3 PUSH"

/// Updates the current value of \`c3\`. Usually, it is used for updating smart contract code in run-time.
/// Note that after execution of this primitive the current code
/// (and the stack of recursive function calls) won't change,
/// but any other function call will use a function from the new code.
fun setTvmRegisterC3(c: continuation): void
    asm "c3 POP"

/// Transforms a slice into a simple ordinary continuation, with \`c.code = s\` and an empty stack and savelist.
@pure
fun transformSliceToContinuation(s: slice): continuation
    asm "BLESS"

/// Moves a variable to the top of the stack.
@pure
fun T.stackMoveToTop(mutate self): void
    asm "NOP"


/**
    Standard TVM exit codes.
    [See docs](https://docs.ton.org/tvm/exit-codes)
 */

/// Standard exit codes of the compute and action phases, reserved by TVM.
/// Exit codes 0 and 1 indicate normal (successful) execution of the compute phase.
/// Exit code 0 indicates normal (successful) execution of the action phase.
/// TON Blockchain reserves exit code values from 0 to 127.
/// For custom errors, use numbers from 128 to 1024 (above is not recommended due to extra asm instructions).
enum VmExitCode {
    NormalTermination = 0           // successful execution of compute or action phase
    AlternativeTermination = 1      // alternative success code; reserved, but never occurs
    StackUnderflow = 2              // not enough stack values for the instruction
    StackOverflow = 3               // too many elements copied into a closure continuation
    IntegerOverflow = 4             // integer does not fit into −2^256 .. 2^256−1, or division by zero
    IntegerOutOfRange = 5           // integer is out of expected range (range check error)
    InvalidOpcode = 6               // instruction or its fift mapping not found
    TypeCheckError = 7              // argument to a primitive is of incorrect value type
    CellOverflow = 8                // writing too much data or refs to a builder
    CellUnderflow = 9               // reading more data or refs than a slice contains
    DictionaryError = 10            // dictionary (hashmap) manipulation error
    UnknownError = 11               // unknown error, may be thrown by user programs
    FatalError = 12                 // thrown by TVM in situations deemed impossible
    OutOfGasError = 13              // not enough gas to finish execution (shown as -14 at runtime)
    VirtualizationError = 14        // reserved, but never thrown
    ActionListInvalid = 32          // action list contains exotic or unparseable cells
    ActionListTooLong = 33          // more than 255 actions queued
    ActionInvalid = 34              // action is invalid or not supported
    InvalidSourceAddress = 35       // invalid source address in outbound message
    InvalidDestAddress = 36         // invalid destination address in outbound message
    NotEnoughToncoin = 37           // not enough Toncoin to send or pay fees
    NotEnoughExtraCurrencies = 38   // not enough extra currencies to send the specified amount
    MessageDoesNotFit = 39          // outbound message does not fit into a cell after rewriting
    CannotProcessMessage = 40       // not enough funds, message too large, or Merkle depth too big
    LibraryReferenceNull = 41       // library reference is null during library change action
    LibraryChangeError = 42         // error during library change action
    LibraryLimitsExceeded = 43      // too many library cells or Merkle tree depth exceeded
    AccountStateSizeExceeded = 50   // account state exceeded limits (config param 43)
    OpcodeDoesNotMatch = 63         // default opcode mismatch for [T.fromCell] and [T.fromSlice]
}
`,

'@fiftlib/Asm.fif':`library TVM_Asm
// simple TVM Assembler
namespace Asm
Asm definitions
"0.4.6" constant asm-fif-version

variable @atend
variable @was-split
false @was-split !
{ "not in asm context" abort } @atend !
{ \`normal eq? not abort"must be terminated by }>" } : @normal?
{ context@ @atend @ 2 { @atend ! context! @normal? } does @atend ! } : @pushatend
{ @pushatend Asm <b } : <{
{ @atend @ execute } : @endblk
{ false @was-split ! \`normal @endblk } : }>
{ }> b> } : }>c
{ }>c <s } : }>s
{ @atend @ 2 { true @was-split ! @atend ! rot b> ref, swap @endblk } does @atend ! <b } : @|
{ @atend @ 3 { @atend ! 2swap rot execute } does @atend ! <b } : @doafter<{
{ over brembits <= } : @havebits
{ rot + -rot + swap } : pair+
{ rot >= -rot <= and } : 2x<=
{ 2 pick brembitrefs 1- 2x<= } : @havebitrefs
{ @havebits ' @| ifnot } : @ensurebits
{ @havebitrefs ' @| ifnot } : @ensurebitrefs
{ rot over @ensurebits -rot u, } : @simpleuop
{ tuck sbitrefs @ensurebitrefs swap s, } : @addop
{ tuck bbitrefs @ensurebitrefs swap b+ } : @addopb
' @addopb : @inline
{ 1 ' @addop does create } : @Defop
{ 1 { <b swap s, swap 8 u, @addopb } does create } : @Defop(8u)
{ 1 { <b swap s, swap 8 i, @addopb } does create } : @Defop(8i)
{ 1 { <b swap s, swap 1- 8 u, @addopb } does create } : @Defop(8u+1)
{ 1 { <b swap s, swap 4 u, @addopb } does create } : @Defop(4u)
{ 1 { <b swap s, swap 12 u, @addopb } does create } : @Defop(12u)
{ 1 { <b swap s, rot 4 u, swap 4 u, @addopb } does create } : @Defop(4u,4u)
{ 1 { <b swap s, swap ref, @addopb } does create } : @Defop(ref)
{ 1 { <b swap s, rot ref, swap ref, @addopb } does create } : @Defop(ref*2)
{ <b 0xef 8 u, swap 12 i, b> } : si()
// x mi ma -- ?
{ rot tuck >= -rot <= and } : @range
{ rot tuck < -rot > or } : @-range
{ @-range abort"Out of range" } : @rangechk
{ dup 0 < over 255 > or abort"Invalid stack register number" si() } : s()
{ si() constant } : @Sreg
-2 @Sreg s(-2)
-1 @Sreg s(-1)
0 @Sreg s0
1 @Sreg s1
2 @Sreg s2
3 @Sreg s3
4 @Sreg s4
5 @Sreg s5
6 @Sreg s6
7 @Sreg s7
8 @Sreg s8
9 @Sreg s9
10 @Sreg s10
11 @Sreg s11
12 @Sreg s12
13 @Sreg s13
14 @Sreg s14
15 @Sreg s15
{ dup 0 < over 7 > or abort"Invalid control register number" <b 0xcc 8 u, swap 4 u, b> } : c()
{ c() constant } : @Creg
0 @Creg c0
1 @Creg c1
2 @Creg c2
3 @Creg c3
4 @Creg c4
5 @Creg c5
7 @Creg c7
{ <s 8 u@+ swap 0xef <> abort"not a stack register" 12 i@+ s> } : @bigsridx
{ @bigsridx dup 16 >= over 0< or abort"stack register s0..s15 expected" } : @sridx
{ rot @bigsridx tuck < -rot tuck > rot or abort"stack register out of range" } : @sridxrange
{ swap @bigsridx + dup 16 >= over 0< or abort"stack register out of range" } : @sridx+
{ <s 8 u@+ 4 u@+ s> swap 0xcc <> over 7 > or over 6 = or abort"not a control register c0..c5 or c7" } : @cridx
{ <s 8 u@ 0xcc = } : @iscr?
{ <b swap s, 1 { swap @sridx 4 u, @addopb } does create } : @Defop(s)
{ <b swap s, 1 { rot @sridx 4 u, swap @sridx 4 u, @addopb } does create } : @Defop(s,s)
{ <b swap s, 1 { swap @cridx 4 u, @addopb } does create } : @Defop(c)
//
// stack manipulation primitives
// (simple stack primitives)
//
x{00} @Defop NOP
x{01} @Defop SWAP
x{0} @Defop(s) XCHG0
{ @bigsridx swap @bigsridx 2dup =
  { 2drop <b }
  { 2dup < { swap } if dup 0=
    { drop dup 16 <
      { <b x{0} s, swap 4 u, }
      { <b x{11} s, swap 8 u, }
      cond
    }
    { over 16 >=
      { tuck 16 >=
        { <b x{11} s, 2 pick 8 u, x{11} s, swap 8 u, x{11} s, swap 8 u, }
	{ <b x{0} s, 2 pick 4 u, x{11} s, swap 8 u, x{0} s, swap 4 u, } cond
      }
      { dup 1 =
        { drop <b x{1} s, swap 4 u, }
        { <b x{10} s, swap 4 u, swap 4 u, }
        cond
      } cond
    } cond
  } cond
  @addopb } : XCHG
x{ED4} @Defop(c) PUSHCTR
x{ED5} @Defop(c) POPCTR
{ dup @iscr?
  ' PUSHCTR
  { @bigsridx dup 16 <
    { <b x{2} s, swap 4 u, }
    { <b x{56} s, swap 8 u,
    } cond
  @addopb
  } cond
} : PUSH
x{20} @Defop DUP
x{21} @Defop OVER
{ dup @iscr?
  ' POPCTR
  { @bigsridx dup 16 <
    { <b x{3} s, swap 4 u, }
    { <b x{57} s, swap 8 u,
    } cond
  @addopb
  } cond
} : POP
x{30} @Defop DROP
x{31} @Defop NIP

// compound stack primitives
{ @sridx rot @sridx rot @sridx swap <b x{4} s, swap 4 u, swap 4 u, swap 4 u, @addopb } : XCHG3
x{50} @Defop(s,s) XCHG2
x{51} @Defop(s,s) XCPU
{ <b x{52} s, rot @sridx 4 u, swap 1 @sridx+ 4 u, @addopb } : PUXC
x{53} @Defop(s,s) PUSH2
{ @sridx rot @sridx rot @sridx swap <b x{540} s, swap 4 u, swap 4 u, swap 4 u, @addopb } : XCHG3_l
{ @sridx rot @sridx rot @sridx swap <b x{541} s, swap 4 u, swap 4 u, swap 4 u, @addopb } : XC2PU
{ 1 @sridx+ rot @sridx rot @sridx swap <b x{542} s, swap 4 u, swap 4 u, swap 4 u, @addopb } : XCPUXC
{ @sridx rot @sridx rot @sridx swap <b x{543} s, swap 4 u, swap 4 u, swap 4 u, @addopb } : XCPU2
{ 1 @sridx+ rot @sridx rot 1 @sridx+ swap <b x{544} s, swap 4 u, swap 4 u, swap 4 u, @addopb } : PUXC2
{ 1 @sridx+ rot @sridx rot 1 @sridx+ swap <b x{545} s, swap 4 u, swap 4 u, swap 4 u, @addopb } : PUXCPU
{ 2 @sridx+ rot @sridx rot 1 @sridx+ swap <b x{546} s, swap 4 u, swap 4 u, swap 4 u, @addopb } : PU2XC
{ @sridx rot @sridx rot @sridx swap <b x{547} s, swap 4 u, swap 4 u, swap 4 u, @addopb } : PUSH3
{ <b x{55} s, rot 1- 4 u, swap 1- 4 u, @addopb } : BLKSWAP
{ dup { 1 swap BLKSWAP } { drop } cond } : ROLL
{ dup { 1 BLKSWAP } { drop } cond } dup : -ROLL : ROLLREV
x{5513} dup @Defop 2ROT @Defop ROT2

// exotic stack primitives
x{58} @Defop ROT
x{59} dup @Defop -ROT @Defop ROTREV
x{5A} dup @Defop 2SWAP @Defop SWAP2
x{5B} dup @Defop 2DROP @Defop DROP2
x{5C} dup @Defop 2DUP @Defop DUP2
x{5D} dup @Defop 2OVER @Defop OVER2
{ <b x{5E} s, rot 2 - 4 u, swap 4 u, @addopb } : REVERSE
{ <b x{5F0} s, swap 4 u, @addopb } : BLKDROP
{ over 0= abort"first argument must be non-zero"
  <b x{5F} s, rot 4 u, swap 4 u, @addopb } : BLKPUSH
x{60} dup @Defop PICK @Defop PUSHX
x{61} @Defop ROLLX
x{62} dup @Defop -ROLLX @Defop ROLLREVX
x{63} @Defop BLKSWX
x{64} @Defop REVX
x{65} @Defop DROPX
x{66} @Defop TUCK
x{67} @Defop XCHGX
x{68} @Defop DEPTH
x{69} @Defop CHKDEPTH
x{6A} @Defop ONLYTOPX
x{6B} @Defop ONLYX
{ over 0= abort"first argument must be non-zero"
  <b x{6C} s, rot 4 u, swap 4 u, @addopb } : BLKDROP2

// null primitives
x{6D} dup @Defop NULL @Defop PUSHNULL
x{6E} @Defop ISNULL
// tuple primitives
x{6F0} @Defop(4u) TUPLE
x{6F00} @Defop NIL
x{6F01} @Defop SINGLE
x{6F02} dup @Defop PAIR @Defop CONS
x{6F03} @Defop TRIPLE
x{6F1} @Defop(4u) INDEX
x{6F10} dup @Defop FIRST @Defop CAR
x{6F11} dup @Defop SECOND @Defop CDR
x{6F12} @Defop THIRD
x{6F2} @Defop(4u) UNTUPLE
x{6F21} @Defop UNSINGLE
x{6F22} dup @Defop UNPAIR @Defop UNCONS
x{6F23} @Defop UNTRIPLE
x{6F3} @Defop(4u) UNPACKFIRST
x{6F30} @Defop CHKTUPLE
x{6F4} @Defop(4u) EXPLODE
x{6F5} @Defop(4u) SETINDEX
x{6F50} @Defop SETFIRST
x{6F51} @Defop SETSECOND
x{6F52} @Defop SETTHIRD
x{6F6} @Defop(4u) INDEXQ
x{6F60} dup @Defop FIRSTQ @Defop CARQ
x{6F61} dup @Defop SECONDQ @Defop CDRQ
x{6F62} @Defop THIRDQ
x{6F7} @Defop(4u) SETINDEXQ
x{6F70} @Defop SETFIRSTQ
x{6F71} @Defop SETSECONDQ
x{6F72} @Defop SETTHIRDQ
x{6F80} @Defop TUPLEVAR
x{6F81} @Defop INDEXVAR
x{6F82} @Defop UNTUPLEVAR
x{6F83} @Defop UNPACKFIRSTVAR
x{6F84} @Defop EXPLODEVAR
x{6F85} @Defop SETINDEXVAR
x{6F86} @Defop INDEXVARQ
x{6F87} @Defop SETINDEXVARQ
x{6F88} @Defop TLEN
x{6F89} @Defop QTLEN
x{6F8A} @Defop ISTUPLE
x{6F8B} @Defop LAST
x{6F8C} dup @Defop TPUSH @Defop COMMA
x{6F8D} @Defop TPOP
x{6FA0} @Defop NULLSWAPIF
x{6FA1} @Defop NULLSWAPIFNOT
x{6FA2} @Defop NULLROTRIF
x{6FA3} @Defop NULLROTRIFNOT
x{6FA4} @Defop NULLSWAPIF2
x{6FA5} @Defop NULLSWAPIFNOT2
x{6FA6} @Defop NULLROTRIF2
x{6FA7} @Defop NULLROTRIFNOT2
{ <b x{6FB} s, rot 2 u, swap 2 u, @addopb } : INDEX2
x{6FB4} @Defop CADR
x{6FB5} @Defop CDDR
{ <b x{6FE_} s, 3 roll 2 u, rot 2 u, swap 2 u, @addopb } : INDEX3
x{6FD4} @Defop CADDR
x{6FD5} @Defop CDDDR

// integer constants
x{70} dup @Defop ZERO @Defop FALSE
x{71} @Defop ONE
x{72} @Defop TWO
x{7A} @Defop TEN
x{7F} @Defop TRUE
{ dup 10 <= over -5 >= and
  { 15 and <b x{7} s, swap 4 u, }
  { dup 8 fits
    { <b x{80} s, swap 8 i, }
    { dup 16 fits
      { <b x{81} s, swap 16 i, }
      { 11 { dup 259 > abort"integer too large" 8 + 2dup fits } until
        <b x{82} s, over 3 >> 2- 5 u, -rot i,
      } cond
    } cond
  } cond
  @addopb } dup : PUSHINT : INT
{ dup 256 = abort"use PUSHNAN instead of 256 PUSHPOW2" <b x{83} s, swap 1- 8 u, @addopb } : PUSHPOW2
x{83FF} @Defop PUSHNAN
{ <b x{84} s, swap 1- 8 u, @addopb } : PUSHPOW2DEC
{ <b x{85} s, swap 1- 8 u, @addopb } : PUSHNEGPOW2
//
// other constants
x{88} @Defop(ref) PUSHREF
x{89} @Defop(ref) PUSHREFSLICE
x{8A} @Defop(ref) PUSHREFCONT
{ 1- dup 0< over 8 >= or abort"invalid slice padding"
  swap 1 1 u, 0 rot u, } : @scomplete
{ tuck sbitrefs swap 26 + swap @havebitrefs not
  { <b rot s, b> PUSHREFSLICE }
  { over sbitrefs 2dup 123 0 2x<=
    { drop tuck 4 + 3 >> swap x{8B} s, over 4 u, 3 roll s,
      -rot 3 << 4 + swap - @scomplete }
    { 2dup 1 >= swap 248 <= and
      { rot x{8C} s, swap 1- 2 u, over 7 + 3 >> tuck 5 u, 3 roll s,
        -rot 3 << 1 + swap - @scomplete }
      { rot x{8D} s, swap 3 u, over 2 + 3 >> tuck 7 u, 3 roll s,
        -rot 3 << 6 + swap - @scomplete
      } cond
    } cond
  } cond
} dup : PUSHSLICE : SLICE
// ( b' -- ? )
{ bbitrefs or 0= } : @cont-empty?
{ bbits 7 and 0= } : @cont-aligned?
// ( b b' -- ? )
{ bbitrefs over 7 and { 2drop drop false } {
  swap 16 + swap @havebitrefs nip
  } cond
} : @cont-fits?
// ( b b' -- ? )
{ bbitrefs over 7 and { 2drop drop false } {
  32 1 pair+ @havebitrefs nip
  } cond
} : @cont-ref-fit?
// ( b b' b'' -- ? )
{ over @cont-aligned? over @cont-aligned? and not { 2drop drop false } {
  bbitrefs rot bbitrefs pair+ swap 32 + swap @havebitrefs nip
  } cond
} : @two-cont-fit?
{ 2dup @cont-fits? not
  { b> PUSHREFCONT }
  { swap over bbitrefs 2dup 120 0 2x<=
    { drop swap x{9} s, swap 3 >> 4 u, swap b+ }
    { rot x{8F_} s, swap 2 u, swap 3 >> 7 u, swap b+ } cond
  } cond
} dup : PUSHCONT : CONT
{ }> PUSHCONT } : }>CONT
{ { @normal? PUSHCONT } @doafter<{ } : CONT:<{

// arithmetic operations
{ 2 { rot dup 8 fits
      { nip <b rot s, swap 8 i, }
      { rot drop <b swap PUSHINT swap s, }
      cond @addopb
    } does create
} : @Defop(8i,alt)
x{A0} @Defop ADD
x{A1} @Defop SUB
x{A2} @Defop SUBR
x{A3} @Defop NEGATE
x{A4} @Defop INC
x{A5} @Defop DEC
x{A6} x{A0} @Defop(8i,alt) ADDCONST
{ negate ADDCONST } : SUBCONST
x{A7} x{A8} @Defop(8i,alt) MULCONST
' ADDCONST : ADDINT
' SUBCONST : SUBINT
' MULCONST : MULINT
x{A8} @Defop MUL

x{A904} @Defop DIV
x{A905} @Defop DIVR
x{A906} @Defop DIVC
x{A908} @Defop MOD
x{A909} @Defop MODR
x{A90A} @Defop MODC
x{A90C} @Defop DIVMOD
x{A90D} @Defop DIVMODR
x{A90E} @Defop DIVMODC
x{A900} @Defop ADDDIVMOD
x{A901} @Defop ADDDIVMODR
x{A902} @Defop ADDDIVMODC

x{A925} @Defop RSHIFTR
x{A926} @Defop RSHIFTC
x{A928} @Defop MODPOW2
x{A929} @Defop MODPOW2R
x{A92A} @Defop MODPOW2C
x{A92C} @Defop RSHIFTMOD
x{A92D} @Defop RSHIFTMODR
x{A92E} @Defop RSHIFTMODC
x{A920} @Defop ADDRSHIFTMOD
x{A921} @Defop ADDRSHIFTMODR
x{A922} @Defop ADDRSHIFTMODC

x{A935} @Defop(8u+1) RSHIFTR#
x{A936} @Defop(8u+1) RSHIFTC#
x{A938} @Defop(8u+1) MODPOW2#
x{A939} @Defop(8u+1) MODPOW2R#
x{A93A} @Defop(8u+1) MODPOW2C#
x{A93C} @Defop(8u+1) RSHIFT#MOD
x{A93D} @Defop(8u+1) RSHIFTR#MOD
x{A93E} @Defop(8u+1) RSHIFTC#MOD
x{A930} @Defop(8u+1) ADDRSHIFT#MOD
x{A931} @Defop(8u+1) ADDRSHIFTR#MOD
x{A932} @Defop(8u+1) ADDRSHIFTC#MOD

x{A984} @Defop MULDIV
x{A985} @Defop MULDIVR
x{A986} @Defop MULDIVC
x{A988} @Defop MULMOD
x{A989} @Defop MULMODR
x{A98A} @Defop MULMODC
x{A98C} @Defop MULDIVMOD
x{A98D} @Defop MULDIVMODR
x{A98E} @Defop MULDIVMODC
x{A980} @Defop MULADDDIVMOD
x{A981} @Defop MULADDDIVMODR
x{A982} @Defop MULADDDIVMODC

x{A9A4} @Defop MULRSHIFT
x{A9A5} @Defop MULRSHIFTR
x{A9A6} @Defop MULRSHIFTC
x{A9A8} @Defop MULMODPOW2
x{A9A9} @Defop MULMODPOW2R
x{A9AA} @Defop MULMODPOW2C
x{A9AC} @Defop MULRSHIFTMOD
x{A9AD} @Defop MULRSHIFTRMOD
x{A9AE} @Defop MULRSHIFTCMOD
x{A9A0} @Defop MULADDRSHIFTMOD
x{A9A1} @Defop MULADDRSHIFTRMOD
x{A9A2} @Defop MULADDRSHIFTCMOD

x{A9B4} @Defop(8u+1) MULRSHIFT#
x{A9B5} @Defop(8u+1) MULRSHIFTR#
x{A9B6} @Defop(8u+1) MULRSHIFTC#
x{A9B8} @Defop(8u+1) MULMODPOW2#
x{A9B9} @Defop(8u+1) MULMODPOW2R#
x{A9BA} @Defop(8u+1) MULMODPOW2C#
x{A9BC} @Defop(8u+1) MULRSHIFT#MOD
x{A9BD} @Defop(8u+1) MULRSHIFTR#MOD
x{A9BE} @Defop(8u+1) MULRSHIFTC#MOD
x{A9B0} @Defop(8u+1) MULADDRSHIFT#MOD
x{A9B1} @Defop(8u+1) MULADDRSHIFTR#MOD
x{A9B2} @Defop(8u+1) MULADDRSHIFTC#MOD

x{A9C4} @Defop LSHIFTDIV
x{A9C5} @Defop LSHIFTDIVR
x{A9C6} @Defop LSHIFTDIVC
x{A9C8} @Defop LSHIFTMOD
x{A9C9} @Defop LSHIFTMODR
x{A9CA} @Defop LSHIFTMODC
x{A9CC} @Defop LSHIFTDIVMOD
x{A9CD} @Defop LSHIFTDIVMODR
x{A9CE} @Defop LSHIFTDIVMODC
x{A9C0} @Defop LSHIFTADDDIVMOD
x{A9C1} @Defop LSHIFTADDDIVMODR
x{A9C2} @Defop LSHIFTADDDIVMODC

x{A9D4} @Defop(8u+1) LSHIFT#DIV
x{A9D5} @Defop(8u+1) LSHIFT#DIVR
x{A9D6} @Defop(8u+1) LSHIFT#DIVC
x{A9D8} @Defop(8u+1) LSHIFT#MOD
x{A9D9} @Defop(8u+1) LSHIFT#MODR
x{A9DA} @Defop(8u+1) LSHIFT#MODC
x{A9DC} @Defop(8u+1) LSHIFT#DIVMOD
x{A9DD} @Defop(8u+1) LSHIFT#DIVMODR
x{A9DE} @Defop(8u+1) LSHIFT#DIVMODC
x{A9D0} @Defop(8u+1) LSHIFT#ADDDIVMOD
x{A9D1} @Defop(8u+1) LSHIFT#ADDDIVMODR
x{A9D2} @Defop(8u+1) LSHIFT#ADDDIVMODC

x{AA} @Defop(8u+1) LSHIFT#
x{AB} @Defop(8u+1) RSHIFT#
x{AC} @Defop LSHIFT
x{AD} @Defop RSHIFT
x{AE} @Defop POW2
x{B0} @Defop AND
x{B1} @Defop OR
x{B2} @Defop XOR
x{B3} @Defop NOT
x{B4} @Defop(8u+1) FITS
x{B400} @Defop CHKBOOL
x{B5} @Defop(8u+1) UFITS
x{B500} @Defop CHKBIT
x{B600} @Defop FITSX
x{B601} @Defop UFITSX
x{B602} @Defop BITSIZE
x{B603} @Defop UBITSIZE
x{B608} @Defop MIN
x{B609} @Defop MAX
x{B60A} dup @Defop MINMAX @Defop INTSORT2
x{B60B} @Defop ABS
x{B7} @Defop QUIET
x{B7A0} @Defop QADD
x{B7A1} @Defop QSUB
x{B7A2} @Defop QSUBR
x{B7A3} @Defop QNEGATE
x{B7A4} @Defop QINC
x{B7A5} @Defop QDEC
x{B7A8} @Defop QMUL

x{B7A904} @Defop QDIV
x{B7A905} @Defop QDIVR
x{B7A906} @Defop QDIVC
x{B7A908} @Defop QMOD
x{B7A909} @Defop QMODR
x{B7A90A} @Defop QMODC
x{B7A90C} @Defop QDIVMOD
x{B7A90D} @Defop QDIVMODR
x{B7A90E} @Defop QDIVMODC
x{B7A900} @Defop QADDDIVMOD
x{B7A901} @Defop QADDDIVMODR
x{B7A902} @Defop QADDDIVMODC

x{B7A925} @Defop QRSHIFTR
x{B7A926} @Defop QRSHIFTC
x{B7A928} @Defop QMODPOW2
x{B7A929} @Defop QMODPOW2R
x{B7A92A} @Defop QMODPOW2C
x{B7A92C} @Defop QRSHIFTMOD
x{B7A92D} @Defop QRSHIFTMODR
x{B7A92E} @Defop QRSHIFTMODC
x{B7A920} @Defop QADDRSHIFTMOD
x{B7A921} @Defop QADDRSHIFTMODR
x{B7A922} @Defop QADDRSHIFTMODC

x{B7A935} @Defop(8u+1) QRSHIFTR#
x{B7A936} @Defop(8u+1) QRSHIFTC#
x{B7A938} @Defop(8u+1) QMODPOW2#
x{B7A939} @Defop(8u+1) QMODPOW2R#
x{B7A93A} @Defop(8u+1) QMODPOW2C#
x{B7A93C} @Defop(8u+1) QRSHIFT#MOD
x{B7A93D} @Defop(8u+1) QRSHIFTR#MOD
x{B7A93E} @Defop(8u+1) QRSHIFTC#MOD
x{B7A930} @Defop(8u+1) QADDRSHIFT#MOD
x{B7A931} @Defop(8u+1) QADDRSHIFTR#MOD
x{B7A932} @Defop(8u+1) QADDRSHIFTC#MOD

x{B7A984} @Defop QMULDIV
x{B7A985} @Defop QMULDIVR
x{B7A986} @Defop QMULDIVC
x{B7A988} @Defop QMULMOD
x{B7A989} @Defop QMULMODR
x{B7A98A} @Defop QMULMODC
x{B7A98C} @Defop QMULDIVMOD
x{B7A98D} @Defop QMULDIVMODR
x{B7A98E} @Defop QMULDIVMODC
x{B7A980} @Defop QMULADDDIVMOD
x{B7A981} @Defop QMULADDDIVMODR
x{B7A982} @Defop QMULADDDIVMODC

x{B7A9A4} @Defop QMULRSHIFT
x{B7A9A5} @Defop QMULRSHIFTR
x{B7A9A6} @Defop QMULRSHIFTC
x{B7A9A8} @Defop QMULMODPOW2
x{B7A9A9} @Defop QMULMODPOW2R
x{B7A9AA} @Defop QMULMODPOW2C
x{B7A9AC} @Defop QMULRSHIFTMOD
x{B7A9AD} @Defop QMULRSHIFTRMOD
x{B7A9AE} @Defop QMULRSHIFTCMOD
x{B7A9A0} @Defop QMULADDRSHIFTMOD
x{B7A9A1} @Defop QMULADDRSHIFTRMOD
x{B7A9A2} @Defop QMULADDRSHIFTCMOD

x{B7A9B4} @Defop(8u+1) QMULRSHIFT#
x{B7A9B5} @Defop(8u+1) QMULRSHIFTR#
x{B7A9B6} @Defop(8u+1) QMULRSHIFTC#
x{B7A9B8} @Defop(8u+1) QMULMODPOW2#
x{B7A9B9} @Defop(8u+1) QMULMODPOW2R#
x{B7A9BA} @Defop(8u+1) QMULMODPOW2C#
x{B7A9BC} @Defop(8u+1) QMULRSHIFT#MOD
x{B7A9BD} @Defop(8u+1) QMULRSHIFTR#MOD
x{B7A9BE} @Defop(8u+1) QMULRSHIFTC#MOD
x{B7A9B0} @Defop(8u+1) QMULADDRSHIFT#MOD
x{B7A9B1} @Defop(8u+1) QMULADDRSHIFTR#MOD
x{B7A9B2} @Defop(8u+1) QMULADDRSHIFTC#MOD

x{B7A9C4} @Defop QLSHIFTDIV
x{B7A9C5} @Defop QLSHIFTDIVR
x{B7A9C6} @Defop QLSHIFTDIVC
x{B7A9C8} @Defop QLSHIFTMOD
x{B7A9C9} @Defop QLSHIFTMODR
x{B7A9CA} @Defop QLSHIFTMODC
x{B7A9CC} @Defop QLSHIFTDIVMOD
x{B7A9CD} @Defop QLSHIFTDIVMODR
x{B7A9CE} @Defop QLSHIFTDIVMODC
x{B7A9C0} @Defop QLSHIFTADDDIVMOD
x{B7A9C1} @Defop QLSHIFTADDDIVMODR
x{B7A9C2} @Defop QLSHIFTADDDIVMODC

x{B7A9D4} @Defop(8u+1) QLSHIFT#DIV
x{B7A9D5} @Defop(8u+1) QLSHIFT#DIVR
x{B7A9D6} @Defop(8u+1) QLSHIFT#DIVC
x{B7A9D8} @Defop(8u+1) QLSHIFT#MOD
x{B7A9D9} @Defop(8u+1) QLSHIFT#MODR
x{B7A9DA} @Defop(8u+1) QLSHIFT#MODC
x{B7A9DC} @Defop(8u+1) QLSHIFT#DIVMOD
x{B7A9DD} @Defop(8u+1) QLSHIFT#DIVMODR
x{B7A9DE} @Defop(8u+1) QLSHIFT#DIVMODC
x{B7A9D0} @Defop(8u+1) QLSHIFT#ADDDIVMOD
x{B7A9D1} @Defop(8u+1) QLSHIFT#ADDDIVMODR
x{B7A9D2} @Defop(8u+1) QLSHIFT#ADDDIVMODC

x{B7AC} @Defop QLSHIFT
x{B7AD} @Defop QRSHIFT
x{B7AE} @Defop QPOW2
x{B7B0} @Defop QAND
x{B7B1} @Defop QOR
x{B7B2} @Defop QXOR
x{B7B3} @Defop QNOT
x{B7B4} @Defop(8u+1) QFITS
x{B7B5} @Defop(8u+1) QUFITS
x{B7B600} @Defop QFITSX
x{B7B601} @Defop QUFITSX

// advanced integer constants
{ 0 { over 1 and 0= } { 1+ swap 2/ swap } while } : pow2decomp
{ dup 8 fits { PUSHINT } {
  dup pow2decomp over 1 = { nip nip PUSHPOW2 } {
  over -1 = { nip nip PUSHNEGPOW2 } {
  dup 20 >= { rot drop -rot PUSHINT swap LSHIFT# } {
  { drop PUSHINT } {
  not pow2decomp swap -1 = { nip PUSHPOW2DEC } {
  drop PUSHINT
  } cond } cond } cond } cond } cond } cond
} dup : PUSHINTX : INTX

// integer comparison
x{B8} @Defop SGN
x{B9} @Defop LESS
x{BA} @Defop EQUAL
x{BB} @Defop LEQ
x{BC} @Defop GREATER
x{BD} @Defop NEQ
x{BE} @Defop GEQ
x{BF} @Defop CMP
x{C0} x{BA} @Defop(8i,alt) EQINT
x{C000} @Defop ISZERO
x{C1} x{B9} @Defop(8i,alt) LESSINT
{ 1+ LESSINT } : LEQINT
x{C100} @Defop ISNEG
x{C101} @Defop ISNPOS
x{C2} x{BC} @Defop(8i,alt) GTINT
{ 1- GTINT } : GEQINT
x{C200} @Defop ISPOS
x{C2FF} @Defop ISNNEG
x{C3} x{BD} @Defop(8i,alt) NEQINT
x{C300} @Defop ISNZERO
x{C4} @Defop ISNAN
x{C5} @Defop CHKNAN

// other comparison
x{C700} @Defop SEMPTY
x{C701} @Defop SDEMPTY
x{C702} @Defop SREMPTY
x{C703} @Defop SDFIRST
x{C704} @Defop SDLEXCMP
x{C705} @Defop SDEQ
x{C708} @Defop SDPFX
x{C709} @Defop SDPFXREV
x{C70A} @Defop SDPPFX
x{C70B} @Defop SDPPFXREV
x{C70C} @Defop SDSFX
x{C70D} @Defop SDSFXREV
x{C70E} @Defop SDPSFX
x{C70F} @Defop SDPSFXREV
x{C710} @Defop SDCNTLEAD0
x{C711} @Defop SDCNTLEAD1
x{C712} @Defop SDCNTTRAIL0
x{C713} @Defop SDCNTTRAIL1

// cell serialization (Builder manipulation primitives)
x{C8} @Defop NEWC
x{C9} @Defop ENDC
x{CA} @Defop(8u+1) STI
x{CB} @Defop(8u+1) STU
x{CC} @Defop STREF
x{CD} dup @Defop STBREFR @Defop ENDCST
x{CE} @Defop STSLICE
x{CF00} @Defop STIX
x{CF01} @Defop STUX
x{CF02} @Defop STIXR
x{CF03} @Defop STUXR
x{CF04} @Defop STIXQ
x{CF05} @Defop STUXQ
x{CF06} @Defop STIXRQ
x{CF07} @Defop STUXRQ
x{CF08} @Defop(8u+1) STI_l
x{CF09} @Defop(8u+1) STU_l
x{CF0A} @Defop(8u+1) STIR
x{CF0B} @Defop(8u+1) STUR
x{CF0C} @Defop(8u+1) STIQ
x{CF0D} @Defop(8u+1) STUQ
x{CF0E} @Defop(8u+1) STIRQ
x{CF0F} @Defop(8u+1) STURQ
x{CF10} @Defop STREF_l
x{CF11} @Defop STBREF
x{CF12} @Defop STSLICE_l
x{CF13} @Defop STB
x{CF14} @Defop STREFR
x{CF15} @Defop STBREFR_l
x{CF16} @Defop STSLICER
x{CF17} dup @Defop STBR @Defop BCONCAT
x{CF18} @Defop STREFQ
x{CF19} @Defop STBREFQ
x{CF1A} @Defop STSLICEQ
x{CF1B} @Defop STBQ
x{CF1C} @Defop STREFRQ
x{CF1D} @Defop STBREFRQ
x{CF1E} @Defop STSLICERQ
x{CF1F} dup @Defop STBRQ @Defop BCONCATQ
x{CF20} @Defop(ref) STREFCONST
{ <b x{CF21} s, rot ref, swap ref, @addopb } : STREF2CONST
x{CF23} @Defop ENDXC
x{CF28} @Defop STILE4
x{CF29} @Defop STULE4
x{CF2A} @Defop STILE8
x{CF2B} @Defop STULE8
x{CF30} @Defop BDEPTH
x{CF31} @Defop BBITS
x{CF32} @Defop BREFS
x{CF33} @Defop BBITREFS
x{CF35} @Defop BREMBITS
x{CF36} @Defop BREMREFS
x{CF37} @Defop BREMBITREFS
x{CF38} @Defop(8u+1) BCHKBITS#
x{CF39} @Defop BCHKBITS
x{CF3A} @Defop BCHKREFS
x{CF3B} @Defop BCHKBITREFS
x{CF3C} @Defop(8u+1) BCHKBITSQ#
x{CF3D} @Defop BCHKBITSQ
x{CF3E} @Defop BCHKREFSQ
x{CF3F} @Defop BCHKBITREFSQ
x{CF40} @Defop STZEROES
x{CF41} @Defop STONES
x{CF42} @Defop STSAME
x{CF50} @Defop BTOS
{ tuck sbitrefs swap 22 + swap @havebitrefs not
  { swap PUSHSLICE STSLICER }
  { over sbitrefs 2dup 57 3 2x<=
    { rot x{CFC_} s, swap 2 u, over 6 + 3 >> tuck 3 u, 3 roll s,
      -rot 3 << 2 + swap - @scomplete }
    { 2drop swap PUSHSLICE STSLICER } cond
  } cond
} : STSLICECONST
x{CF81} @Defop STZERO
x{CF83} @Defop STONE

// cell deserialization (CellSlice primitives)
x{D0} @Defop CTOS
x{D1} @Defop ENDS
x{D2} @Defop(8u+1) LDI
x{D3} @Defop(8u+1) LDU
x{D4} @Defop LDREF
x{D5} @Defop LDREFRTOS
x{D6} @Defop(8u+1) LDSLICE
x{D700} @Defop LDIX
x{D701} @Defop LDUX
x{D702} @Defop PLDIX
x{D703} @Defop PLDUX
x{D704} @Defop LDIXQ
x{D705} @Defop LDUXQ
x{D706} @Defop PLDIXQ
x{D707} @Defop PLDUXQ
x{D708} @Defop(8u+1) LDI_l
x{D709} @Defop(8u+1) LDU_l
x{D70A} @Defop(8u+1) PLDI
x{D70B} @Defop(8u+1) PLDU
x{D70C} @Defop(8u+1) LDIQ
x{D70D} @Defop(8u+1) LDUQ
x{D70E} @Defop(8u+1) PLDIQ
x{D70F} @Defop(8u+1) PLDUQ
{ dup 31 and abort"argument must be a multiple of 32" 5 >> 1-
  <b x{D714_} s, swap 3 u, @addopb
} : PLDUZ
x{D718} @Defop LDSLICEX
x{D719} @Defop PLDSLICEX
x{D71A} @Defop LDSLICEXQ
x{D71B} @Defop PLDSLICEXQ
x{D71C} @Defop(8u+1) LDSLICE_l
x{D71D} @Defop(8u+1) PLDSLICE
x{D71E} @Defop(8u+1) LDSLICEQ
x{D71F} @Defop(8u+1) PLDSLICEQ
x{D720} @Defop SDCUTFIRST
x{D721} @Defop SDSKIPFIRST
x{D722} @Defop SDCUTLAST
x{D723} @Defop SDSKIPLAST
x{D724} @Defop SDSUBSTR
x{D726} @Defop SDBEGINSX
x{D727} @Defop SDBEGINSXQ

/*
if (rembits >= opcodebits) or (rembits <= 18)
  just write
else
  PUSHSLICE SDBEGINSXQ

---
18 - size of PUSHREFSLICE
*/

{ tuck sbits tuck 5 + 3 >> swap x{D72A_} s, over 7 u, 3 roll s,
  -rot 3 << 3 + swap - @scomplete } : SDBEGINS:imm
{ tuck sbitrefs abort"no references allowed in slice" // slice builder slicebits
  5 + 3 >> 8 * 3 + 21 + // slice builder opcodebits ((sbits + 5) // 8) * 8 + 3 + 21
  over brembits <= // total op bits <= remaining free bits in slice
  { <b rot SDBEGINS:imm @addopb }
  { swap PUSHSLICE SDBEGINSX }
  cond
} : SDBEGINS

{ tuck sbits tuck 5 + 3 >> swap x{D72E_} s, over 7 u, 3 roll s,
  -rot 3 << 3 + swap - @scomplete } : SDBEGINSQ:imm
{ tuck sbitrefs abort"no references allowed in slice" // slice builder slicebits
  5 + 3 >> 8 * 3 + 21 + // slice builder opcodebits
  over brembits <= // total op bits <= remaining free bits in slice
  { <b rot SDBEGINSQ:imm @addopb }
  { swap PUSHSLICE SDBEGINSXQ }
  cond
} : SDBEGINSQ

x{D730} @Defop SCUTFIRST
x{D731} @Defop SSKIPFIRST
x{D732} @Defop SCUTLAST
x{D733} @Defop SSKIPLAST
x{D734} @Defop SUBSLICE
x{D736} @Defop SPLIT
x{D737} @Defop SPLITQ
x{D739} @Defop XCTOS
x{D73A} @Defop XLOAD
x{D73B} @Defop XLOADQ
x{D741} @Defop SCHKBITS
x{D742} @Defop SCHKREFS
x{D743} @Defop SCHKBITREFS
x{D745} @Defop SCHKBITSQ
x{D746} @Defop SCHKREFSQ
x{D747} @Defop SCHKBITREFSQ
x{D748} @Defop PLDREFVAR
x{D749} @Defop SBITS
x{D74A} @Defop SREFS
x{D74B} @Defop SBITREFS
{ <b x{D74E_} s, swap 2 u, @addopb } : PLDREFIDX
x{D74C} @Defop PLDREF
x{D750} @Defop LDILE4
x{D751} @Defop LDULE4
x{D752} @Defop LDILE8
x{D753} @Defop LDULE8
x{D754} @Defop PLDILE4
x{D755} @Defop PLDULE4
x{D756} @Defop PLDILE8
x{D757} @Defop PLDULE8
x{D758} @Defop LDILE4Q
x{D759} @Defop LDULE4Q
x{D75A} @Defop LDILE8Q
x{D75B} @Defop LDULE8Q
x{D75C} @Defop PLDILE4Q
x{D75D} @Defop PLDULE4Q
x{D75E} @Defop PLDILE8Q
x{D75F} @Defop PLDULE8Q
x{D760} @Defop LDZEROES
x{D761} @Defop LDONES
x{D762} @Defop LDSAME
x{D764} @Defop SDEPTH
x{D765} @Defop CDEPTH
x{D766} @Defop CLEVEL
x{D767} @Defop CLEVELMASK
{ <b x{D76A_} s, swap 2 u, @addopb } : CHASHI
{ <b x{D76E_} s, swap 2 u, @addopb } : CDEPTHI
x{D770} @Defop CHASHIX
x{D771} @Defop CDEPTHIX
//
// continuation / flow control primitives
x{D8} dup @Defop EXECUTE @Defop CALLX
x{D9} @Defop JMPX
{ dup 1+
  { <b x{DA} s, rot 4 u, swap 4 u, }
  { drop <b x{DB0} s, swap 4 u,
  } cond @addopb
} : CALLXARGS
x{DB1} @Defop(4u) JMPXARGS
x{DB2} @Defop(4u) RETARGS
x{DB30} dup @Defop RET @Defop RETTRUE
x{DB31} dup @Defop RETALT @Defop RETFALSE
x{DB32} dup @Defop BRANCH @Defop RETBOOL
x{DB34} @Defop CALLCC
x{DB35} @Defop JMPXDATA
{ dup 1+ 0= { 16 + } if
  <b x{DB36} s, rot 4 u, swap 4 u, @addopb
} : CALLCCARGS
x{DB38} @Defop CALLXVARARGS
x{DB39} @Defop RETVARARGS
x{DB3A} @Defop JMPXVARARGS
x{DB3B} @Defop CALLCCVARARGS
x{DB3C} @Defop(ref) CALLREF
x{DB3D} @Defop(ref) JMPREF
x{DB3E} @Defop(ref) JMPREFDATA
x{DB3F} @Defop RETDATA
x{DB4} @Defop(12u) RUNVM
x{DB50} @Defop RUNVMX
// conditional and iterated execution primitives
x{DC} @Defop IFRET
x{DD} @Defop IFNOTRET
x{DE} @Defop IF
x{DF} @Defop IFNOT
' IFNOTRET : IF:
' IFRET : IFNOT:
x{E0} @Defop IFJMP
x{E1} @Defop IFNOTJMP
x{E2} @Defop IFELSE

x{E300} @Defop(ref) IFREF
x{E301} @Defop(ref) IFNOTREF
x{E302} @Defop(ref) IFJMPREF
x{E303} @Defop(ref) IFNOTJMPREF
x{E30D} @Defop(ref) IFREFELSE
x{E30E} @Defop(ref) IFELSEREF
x{E30F} @Defop(ref*2) IFREFELSEREF

{ 16 1 @havebitrefs nip } : @refop-fits?
// b b1 [e0 e1 e2] -- b'
{ -rot dup @cont-empty? { drop swap 0 } {
  2dup @cont-fits? { rot 1 } {
  over @refop-fits? { b> rot 2 } {
  swap @| swap 2dup @cont-fits? { rot 1 } {
  b> rot 2
  } cond } cond } cond } cond
  [] execute
} : @run-cont-op
{ triple 1 ' @run-cont-op does create } : @def-cont-op
{ DROP } { PUSHCONT IF } { IFREF } @def-cont-op IF-cont
{ IFRET } { PUSHCONT IFJMP } { IFJMPREF } @def-cont-op IFJMP-cont
{ DROP } { PUSHCONT IFNOT } { IFNOTREF } @def-cont-op IFNOT-cont
{ IFNOTRET } { PUSHCONT IFNOTJMP } { IFNOTJMPREF } @def-cont-op IFNOTJMP-cont
{ dup 2over rot } : 3dup

recursive IFELSE-cont2 {
  dup @cont-empty? { drop IF-cont } {
  over @cont-empty? { nip IFNOT-cont } {
  3dup @two-cont-fit? { -rot PUSHCONT swap PUSHCONT IFELSE } {
  3dup nip @cont-ref-fit? { rot swap PUSHCONT swap b> IFREFELSE } {
  3dup drop @cont-ref-fit? { -rot PUSHCONT swap b> IFELSEREF } {
  rot 32 2 @havebitrefs { rot b> rot b> IFREFELSEREF } {
  @| -rot IFELSE-cont2
  } cond } cond } cond } cond } cond } cond
} swap !

{ }> IF-cont } : }>IF
{ }> IFNOT-cont } : }>IFNOT
{ }> IFJMP-cont } : }>IFJMP
{ }> IFNOTJMP-cont } : }>IFNOTJMP
{ { @normal? IFJMP-cont } @doafter<{ } : IFJMP:<{
{ { @normal? IFNOTJMP-cont } @doafter<{ } : IFNOTJMP:<{
{ \`else @endblk } : }>ELSE<{
{ \`else: @endblk } : }>ELSE:
{ 1 { swap @normal? swap IFELSE-cont2 } does @doafter<{ } : @doifelse
{ 1 { swap @normal? IFELSE-cont2 } does @doafter<{ } : @doifnotelse
{
  { dup \`else eq?
    { drop @doifelse }
    { dup \`else: eq?
      { drop IFJMP-cont }
      { @normal? IF-cont
      } cond
    } cond
  } @doafter<{
} : IF:<{
{
  { dup \`else eq?
    { drop @doifnotelse }
    { dup \`else: eq?
      { drop IFNOTJMP-cont }
      { @normal? IFNOT-cont
      } cond
    } cond
  } @doafter<{
} : IFNOT:<{

x{E304} @Defop CONDSEL
x{E305} @Defop CONDSELCHK
x{E308} @Defop IFRETALT
x{E309} @Defop IFNOTRETALT
{ <b x{E39_} s, swap 5 u, @addopb } : IFBITJMP
{ <b x{E3B_} s, swap 5 u, @addopb } : IFNBITJMP
{ <b x{E3D_} s, swap 5 u, swap ref, @addopb } : IFBITJMPREF
{ <b x{E3F_} s, swap 5 u, swap ref, @addopb } : IFNBITJMPREF

x{E4} @Defop REPEAT
x{E5} dup @Defop REPEATEND @Defop REPEAT:
x{E6} @Defop UNTIL
x{E7} dup @Defop UNTILEND @Defop UNTIL:
x{E8} @Defop WHILE
x{E9} @Defop WHILEEND
x{EA} @Defop AGAIN
x{EB} dup @Defop AGAINEND @Defop AGAIN:

{ \`do @endblk } : }>DO<{
{ \`do: @endblk } : }>DO:
{ }> PUSHCONT REPEAT } : }>REPEAT
{ { @normal? PUSHCONT REPEAT } @doafter<{ } : REPEAT:<{
{ }> PUSHCONT UNTIL } : }>UNTIL
{ { @normal? PUSHCONT UNTIL } @doafter<{ } : UNTIL:<{
{ PUSHCONT { @normal? PUSHCONT WHILE } @doafter<{ } : @dowhile
{
  { dup \`do eq?
    { drop @dowhile }
    { \`do: eq? not abort"\`}>DO<{\` expected" PUSHCONT WHILEEND
    } cond
  } @doafter<{
} : WHILE:<{
{ }> PUSHCONT AGAIN } : }>AGAIN
{ { @normal? PUSHCONT AGAIN } @doafter<{ } : AGAIN:<{

x{E314} @Defop REPEATBRK
x{E315} @Defop REPEATENDBRK
x{E316} @Defop UNTILBRK
x{E317} dup @Defop UNTILENDBRK @Defop UNTILBRK:
x{E318} @Defop WHILEBRK
x{E319} @Defop WHILEENDBRK
x{E31A} @Defop AGAINBRK
x{E31B} dup @Defop AGAINENDBRK @Defop AGAINBRK:

{ }> PUSHCONT REPEATBRK } : }>REPEATBRK
{ { @normal? PUSHCONT REPEATBRK } @doafter<{ } : REPEATBRK:<{
{ }> PUSHCONT UNTILBRK } : }>UNTILBRK
{ { @normal? PUSHCONT UNTILBRK } @doafter<{ } : UNTILBRK:<{
{ PUSHCONT { @normal? PUSHCONT WHILEBRK } @doafter<{ } : @dowhile
{
  { dup \`do eq?
    { drop @dowhile }
    { \`do: eq? not abort"\`}>DO<{\` expected" PUSHCONT WHILEENDBRK
    } cond
  } @doafter<{
} : WHILEBRK:<{
{ }> PUSHCONT AGAINBRK } : }>AGAINBRK
{ { @normal? PUSHCONT AGAINBRK } @doafter<{ } : AGAINBRK:<{


//
// continuation stack manipulation and continuation creation
//
{ <b x{EC} s, rot 4 u, swap dup 1+ { 16 + } ifnot 4 u, @addopb } : SETCONTARGS
{ 0 swap SETCONTARGS } : SETNUMARGS
x{ED0} @Defop(4u) RETURNARGS
x{ED10} @Defop RETURNVARARGS
x{ED11} @Defop SETCONTVARARGS
x{ED12} @Defop SETNUMVARARGS
x{ED1E} @Defop BLESS
x{ED1F} @Defop BLESSVARARGS
{ <b x{EE} s, rot 4 u, swap dup 1+ { 16 + } ifnot 4 u, @addopb } : BLESSARGS
{ 0 swap BLESSARGS } : BLESSNUMARGS
//
// control register and continuation savelist manipulation
// x{ED4} Defop(c) PUSHCTR
// x{ED5} Defop(c) POPCTR
{ c4 PUSHCTR } : PUSHROOT
{ c4 POPCTR } : POPROOT
x{ED6} dup @Defop(c) SETCONTCTR @Defop(c) SETCONT
x{ED7} @Defop(c) SETRETCTR
x{ED8} @Defop(c) SETALTCTR
x{ED9} dup @Defop(c) POPSAVE @Defop(c) POPCTRSAVE
x{EDA} dup @Defop(c) SAVE @Defop(c) SAVECTR
x{EDB} dup @Defop(c) SAVEALT @Defop(c) SAVEALTCTR
x{EDC} dup @Defop(c) SAVEBOTH @Defop(c) SAVEBOTHCTR
x{EDE0} @Defop PUSHCTRX
x{EDE1} @Defop POPCTRX
x{EDE2} @Defop SETCONTCTRX
x{EDE3} @Defop(8u) SETCONTCTRMANY
x{EDE3} @Defop(8u) SETCONTMANY
x{EDE4} @Defop SETCONTCTRMANYX
x{EDE4} @Defop SETCONTMANYX
x{EDF0} dup @Defop BOOLAND @Defop COMPOS
x{EDF1} dup @Defop BOOLOR @Defop COMPOSALT
x{EDF2} @Defop COMPOSBOTH
x{EDF3} @Defop ATEXIT
{ }> PUSHCONT ATEXIT } : }>ATEXIT
{ { @normal? PUSHCONT ATEXIT } @doafter<{ } : ATEXIT:<{
x{EDF4} @Defop ATEXITALT
{ }> PUSHCONT ATEXITALT } : }>ATEXITALT
{ { @normal? PUSHCONT ATEXITALT } @doafter<{ } : ATEXITALT:<{
x{EDF5} @Defop SETEXITALT
{ }> PUSHCONT SETEXITALT } : }>SETEXITALT
{ { @normal? PUSHCONT SETEXITALT } @doafter<{ } : SETEXITALT:<{
x{EDF6} @Defop THENRET
x{EDF7} @Defop THENRETALT
x{EDF8} @Defop INVERT
x{EDF9} @Defop BOOLEVAL
x{EDFA} @Defop SAMEALT
x{EDFB} @Defop SAMEALTSAVE
// x{EE} is BLESSARGS
//
// dictionary subroutine call/jump primitives
{ c3 PUSH EXECUTE } : CALLVAR
{ c3 PUSH JMPX } : JMPVAR
{ c3 PUSH } : PREPAREVAR
{ dup 14 ufits {
    dup 8 ufits {
      <b x{F0} s, swap 8 u, } {
      <b x{F12_} s, swap 14 u,
    } cond @addopb } {
    PUSHINT CALLVAR
  } cond
} dup : CALL : CALLDICT
{ dup 14 ufits
  { <b x{F16_} s, swap 14 u, @addopb }
  { PUSHINT JMPVAR } cond
} dup : JMP : JMPDICT
{ dup 14 ufits
  { <b x{F1A_} s, swap 14 u, @addopb }
  { PUSHINT PREPAREVAR } cond
} dup : PREPARE : PREPAREDICT
//
// inline support
{ dup sbits
   { @addop }
   {
    dup srefs //
      { ref@ CALLREF }
      { drop }
      cond
   }
   cond
} : INLINE
//
// throwing and handling exceptions
{ dup 6 ufits
  { <b x{F22_} s, swap 6 u, }
  { <b x{F2C4_} s, swap 11 u,
  } cond
@addopb } : THROW
{ dup 6 ufits
  { <b x{F26_} s, swap 6 u, }
  { <b x{F2D4_} s, swap 11 u,
  } cond
@addopb } : THROWIF
{ dup 6 ufits
  { <b x{F2A_} s, swap 6 u, }
  { <b x{F2E4_} s, swap 11 u,
  } cond
@addopb } : THROWIFNOT
{ <b x{F2CC_} s, swap 11 u, @addopb } : THROWARG
{ <b x{F2DC_} s, swap 11 u, @addopb } : THROWARGIF
{ <b x{F2EC_} s, swap 11 u, @addopb } : THROWARGIFNOT
x{F2F0} @Defop THROWANY
x{F2F1} @Defop THROWARGANY
x{F2F2} @Defop THROWANYIF
x{F2F3} @Defop THROWARGANYIF
x{F2F4} @Defop THROWANYIFNOT
x{F2F5} @Defop THROWARGANYIFNOT
x{F2FF} @Defop TRY
x{F3} @Defop(4u,4u) TRYARGS
{ \`catch @endblk } : }>CATCH<{
{ PUSHCONT { @normal? PUSHCONT TRY } @doafter<{ } : @trycatch
{
  { \`catch eq? not abort"\`}>CATCH<{\` expected" @trycatch
  } @doafter<{
} : TRY:<{
//
// dictionary manipulation
' NULL : NEWDICT
' ISNULL : DICTEMPTY
' STSLICE : STDICTS
x{F400} dup @Defop STDICT @Defop STOPTREF
x{F401} dup @Defop SKIPDICT @Defop SKIPOPTREF
x{F402} @Defop LDDICTS
x{F403} @Defop PLDDICTS
x{F404} dup @Defop LDDICT @Defop LDOPTREF
x{F405} dup @Defop PLDDICT @Defop PLDOPTREF
x{F406} @Defop LDDICTQ
x{F407} @Defop PLDDICTQ

x{F40A} @Defop DICTGET
x{F40B} @Defop DICTGETREF
x{F40C} @Defop DICTIGET
x{F40D} @Defop DICTIGETREF
x{F40E} @Defop DICTUGET
x{F40F} @Defop DICTUGETREF

x{F412} @Defop DICTSET
x{F413} @Defop DICTSETREF
x{F414} @Defop DICTISET
x{F415} @Defop DICTISETREF
x{F416} @Defop DICTUSET
x{F417} @Defop DICTUSETREF
x{F41A} @Defop DICTSETGET
x{F41B} @Defop DICTSETGETREF
x{F41C} @Defop DICTISETGET
x{F41D} @Defop DICTISETGETREF
x{F41E} @Defop DICTUSETGET
x{F41F} @Defop DICTUSETGETREF

x{F422} @Defop DICTREPLACE
x{F423} @Defop DICTREPLACEREF
x{F424} @Defop DICTIREPLACE
x{F425} @Defop DICTIREPLACEREF
x{F426} @Defop DICTUREPLACE
x{F427} @Defop DICTUREPLACEREF
x{F42A} @Defop DICTREPLACEGET
x{F42B} @Defop DICTREPLACEGETREF
x{F42C} @Defop DICTIREPLACEGET
x{F42D} @Defop DICTIREPLACEGETREF
x{F42E} @Defop DICTUREPLACEGET
x{F42F} @Defop DICTUREPLACEGETREF

x{F432} @Defop DICTADD
x{F433} @Defop DICTADDREF
x{F434} @Defop DICTIADD
x{F435} @Defop DICTIADDREF
x{F436} @Defop DICTUADD
x{F437} @Defop DICTUADDREF
x{F43A} @Defop DICTADDGET
x{F43B} @Defop DICTADDGETREF
x{F43C} @Defop DICTIADDGET
x{F43D} @Defop DICTIADDGETREF
x{F43E} @Defop DICTUADDGET
x{F43F} @Defop DICTUADDGETREF

x{F441} @Defop DICTSETB
x{F442} @Defop DICTISETB
x{F443} @Defop DICTUSETB
x{F445} @Defop DICTSETGETB
x{F446} @Defop DICTISETGETB
x{F447} @Defop DICTUSETGETB

x{F449} @Defop DICTREPLACEB
x{F44A} @Defop DICTIREPLACEB
x{F44B} @Defop DICTUREPLACEB
x{F44D} @Defop DICTREPLACEGETB
x{F44E} @Defop DICTIREPLACEGETB
x{F44F} @Defop DICTUREPLACEGETB

x{F451} @Defop DICTADDB
x{F452} @Defop DICTIADDB
x{F453} @Defop DICTUADDB
x{F455} @Defop DICTADDGETB
x{F456} @Defop DICTIADDGETB
x{F457} @Defop DICTUADDGETB

x{F459} @Defop DICTDEL
x{F45A} @Defop DICTIDEL
x{F45B} @Defop DICTUDEL

x{F462} @Defop DICTDELGET
x{F463} @Defop DICTDELGETREF
x{F464} @Defop DICTIDELGET
x{F465} @Defop DICTIDELGETREF
x{F466} @Defop DICTUDELGET
x{F467} @Defop DICTUDELGETREF

x{F469} @Defop DICTGETOPTREF
x{F46A} @Defop DICTIGETOPTREF
x{F46B} @Defop DICTUGETOPTREF
x{F46D} @Defop DICTSETGETOPTREF
x{F46E} @Defop DICTISETGETOPTREF
x{F46F} @Defop DICTUSETGETOPTREF

x{F470} @Defop PFXDICTSET
x{F471} @Defop PFXDICTREPLACE
x{F472} @Defop PFXDICTADD
x{F473} @Defop PFXDICTDEL

x{F474} @Defop DICTGETNEXT
x{F475} @Defop DICTGETNEXTEQ
x{F476} @Defop DICTGETPREV
x{F477} @Defop DICTGETPREVEQ
x{F478} @Defop DICTIGETNEXT
x{F479} @Defop DICTIGETNEXTEQ
x{F47A} @Defop DICTIGETPREV
x{F47B} @Defop DICTIGETPREVEQ
x{F47C} @Defop DICTUGETNEXT
x{F47D} @Defop DICTUGETNEXTEQ
x{F47E} @Defop DICTUGETPREV
x{F47F} @Defop DICTUGETPREVEQ

x{F482} @Defop DICTMIN
x{F483} @Defop DICTMINREF
x{F484} @Defop DICTIMIN
x{F485} @Defop DICTIMINREF
x{F486} @Defop DICTUMIN
x{F487} @Defop DICTUMINREF
x{F48A} @Defop DICTMAX
x{F48B} @Defop DICTMAXREF
x{F48C} @Defop DICTIMAX
x{F48D} @Defop DICTIMAXREF
x{F48E} @Defop DICTUMAX
x{F48F} @Defop DICTUMAXREF

x{F492} @Defop DICTREMMIN
x{F493} @Defop DICTREMMINREF
x{F494} @Defop DICTIREMMIN
x{F495} @Defop DICTIREMMINREF
x{F496} @Defop DICTUREMMIN
x{F497} @Defop DICTUREMMINREF
x{F49A} @Defop DICTREMMAX
x{F49B} @Defop DICTREMMAXREF
x{F49C} @Defop DICTIREMMAX
x{F49D} @Defop DICTIREMMAXREF
x{F49E} @Defop DICTUREMMAX
x{F49F} @Defop DICTUREMMAXREF

x{F4A0} @Defop DICTIGETJMP
x{F4A1} @Defop DICTUGETJMP
x{F4A2} @Defop DICTIGETEXEC
x{F4A3} @Defop DICTUGETEXEC
{ dup sbitrefs tuck 1 > swap 1 <> or abort"not a dictionary" swap 1 u@ over <> abort"not a dictionary" } : @chkdicts
{ dup null? tuck { <s } ifnot drop not } : @chkdict
{ over @chkdict
  { swap <b x{F4A6_} s, swap ref, swap 10 u, @addopb }
  { nip swap NEWDICT swap PUSHINT }
  cond
} : DICTPUSHCONST
x{F4A8} @Defop PFXDICTGETQ
x{F4A9} @Defop PFXDICTGET
x{F4AA} @Defop PFXDICTGETJMP
x{F4AB} @Defop PFXDICTGETEXEC
{ over @chkdict
  { swap <b x{F4AE_} s, swap ref, swap 10 u, @addopb
  } if
} dup : PFXDICTCONSTGETJMP : PFXDICTSWITCH

x{F4B1} @Defop SUBDICTGET
x{F4B2} @Defop SUBDICTIGET
x{F4B3} @Defop SUBDICTUGET
x{F4B5} @Defop SUBDICTRPGET
x{F4B6} @Defop SUBDICTIRPGET
x{F4B7} @Defop SUBDICTURPGET

x{F4BC} @Defop DICTIGETJMPZ
x{F4BD} @Defop DICTUGETJMPZ
x{F4BE} @Defop DICTIGETEXECZ
x{F4BF} @Defop DICTUGETEXECZ

//
// blockchain-specific primitives

x{F800} @Defop ACCEPT
x{F801} @Defop SETGASLIMIT
x{F807} @Defop GASCONSUMED
x{F80F} @Defop COMMIT

x{F810} @Defop RANDU256
x{F811} @Defop RAND
x{F814} @Defop SETRAND
x{F815} dup @Defop ADDRAND @Defop RANDOMIZE

x{F82} @Defop(4u) GETPARAM
x{F881} @Defop(8u) GETPARAMLONG
x{F823} @Defop NOW
x{F824} @Defop BLOCKLT
x{F825} @Defop LTIME
x{F826} @Defop RANDSEED
x{F827} @Defop BALANCE
x{F828} @Defop MYADDR
x{F829} @Defop CONFIGROOT
x{F82A} @Defop MYCODE
x{F82B} @Defop INCOMINGVALUE
x{F82C} @Defop STORAGEFEES
x{F82D} @Defop PREVBLOCKSINFOTUPLE
x{F82E} @Defop UNPACKEDCONFIGTUPLE
x{F82F} @Defop DUEPAYMENT
x{F88111} @Defop INMSGPARAMS

x{F830} @Defop CONFIGDICT
x{F832} @Defop CONFIGPARAM
x{F833} @Defop CONFIGOPTPARAM
x{F83400} @Defop PREVMCBLOCKS
x{F83401} @Defop PREVKEYBLOCK
x{F83402} @Defop PREVMCBLOCKS_100
x{F835} @Defop GLOBALID
x{F836} @Defop GETGASFEE
x{F837} @Defop GETSTORAGEFEE
x{F838} @Defop GETFORWARDFEE
x{F839} @Defop GETPRECOMPILEDGAS
x{F83A} @Defop GETORIGINALFWDFEE
x{F83B} @Defop GETGASFEESIMPLE
x{F83C} @Defop GETFORWARDFEESIMPLE

x{F89} @Defop(4u) INMSGPARAM
x{F890} @Defop INMSG_BOUNCE
x{F891} @Defop INMSG_BOUNCED
x{F892} @Defop INMSG_SRC
x{F893} @Defop INMSG_FWDFEE
x{F894} @Defop INMSG_LT
x{F895} @Defop INMSG_UTIME
x{F896} @Defop INMSG_ORIGVALUE
x{F897} @Defop INMSG_VALUE
x{F898} @Defop INMSG_VALUEEXTRA
x{F899} @Defop INMSG_STATEINIT

x{F840} @Defop GETGLOBVAR
{ dup 1 31 @rangechk <b x{F85_} s, swap 5 u, @addopb } : GETGLOB
x{F860} @Defop SETGLOBVAR
{ dup 1 31 @rangechk <b x{F87_} s, swap 5 u, @addopb } : SETGLOB

x{F880} @Defop GETEXTRABALANCE

x{F900} @Defop HASHCU
x{F901} @Defop HASHSU
x{F902} @Defop SHA256U

x{F904} @Defop(8u) HASHEXT
x{F90400} @Defop HASHEXT_SHA256
x{F90401} @Defop HASHEXT_SHA512
x{F90402} @Defop HASHEXT_BLAKE2B
x{F90403} @Defop HASHEXT_KECCAK256
x{F90404} @Defop HASHEXT_KECCAK512
x{F905} @Defop(8u) HASHEXTR
x{F90500} @Defop HASHEXTR_SHA256
x{F90501} @Defop HASHEXTR_SHA512
x{F90502} @Defop HASHEXTR_BLAKE2B
x{F90503} @Defop HASHEXTR_KECCAK256
x{F90504} @Defop HASHEXTR_KECCAK512
x{F906} @Defop(8u) HASHEXTA
x{F90600} @Defop HASHEXTA_SHA256
x{F90601} @Defop HASHEXTA_SHA512
x{F90602} @Defop HASHEXTA_BLAKE2B
x{F90603} @Defop HASHEXTA_KECCAK256
x{F90604} @Defop HASHEXTA_KECCAK512
x{F907} @Defop(8u) HASHEXTAR
x{F90700} @Defop HASHEXTAR_SHA256
x{F90701} @Defop HASHEXTAR_SHA512
x{F90702} @Defop HASHEXTAR_BLAKE2B
x{F90703} @Defop HASHEXTAR_KECCAK256
x{F90704} @Defop HASHEXTAR_KECCAK512

x{F910} @Defop CHKSIGNU
x{F911} @Defop CHKSIGNS
x{F912} @Defop ECRECOVER
x{F913} @Defop SECP256K1_XONLY_PUBKEY_TWEAK_ADD
x{F914} @Defop P256_CHKSIGNU
x{F915} @Defop P256_CHKSIGNS
x{F916} @Defop HASHBU

x{F920} @Defop RIST255_FROMHASH
x{F921} @Defop RIST255_VALIDATE
x{F922} @Defop RIST255_ADD
x{F923} @Defop RIST255_SUB
x{F924} @Defop RIST255_MUL
x{F925} @Defop RIST255_MULBASE
x{F926} @Defop RIST255_PUSHL

x{B7F921} @Defop RIST255_QVALIDATE
x{B7F922} @Defop RIST255_QADD
x{B7F923} @Defop RIST255_QSUB
x{B7F924} @Defop RIST255_QMUL
x{B7F925} @Defop RIST255_QMULBASE

x{F93000} @Defop BLS_VERIFY
x{F93001} @Defop BLS_AGGREGATE
x{F93002} @Defop BLS_FASTAGGREGATEVERIFY
x{F93003} @Defop BLS_AGGREGATEVERIFY

x{F93010} @Defop BLS_G1_ADD
x{F93011} @Defop BLS_G1_SUB
x{F93012} @Defop BLS_G1_NEG
x{F93013} @Defop BLS_G1_MUL
x{F93014} @Defop BLS_G1_MULTIEXP
x{F93015} @Defop BLS_G1_ZERO
x{F93016} @Defop BLS_MAP_TO_G1
x{F93017} @Defop BLS_G1_INGROUP
x{F93018} @Defop BLS_G1_ISZERO

x{F93020} @Defop BLS_G2_ADD
x{F93021} @Defop BLS_G2_SUB
x{F93022} @Defop BLS_G2_NEG
x{F93023} @Defop BLS_G2_MUL
x{F93024} @Defop BLS_G2_MULTIEXP
x{F93025} @Defop BLS_G2_ZERO
x{F93026} @Defop BLS_MAP_TO_G2
x{F93027} @Defop BLS_G2_INGROUP
x{F93028} @Defop BLS_G2_ISZERO

x{F93030} @Defop BLS_PAIRING
x{F93031} @Defop BLS_PUSHR

x{F940} @Defop CDATASIZEQ
x{F941} @Defop CDATASIZE
x{F942} @Defop SDATASIZEQ
x{F943} @Defop SDATASIZE

x{FA00} dup @Defop LDGRAMS @Defop LDVARUINT16
x{FA01} @Defop LDVARINT16
x{FA02} dup @Defop STGRAMS @Defop STVARUINT16
x{FA03} @Defop STVARINT16

x{FA04} @Defop LDVARUINT32 // (s -- x s')
x{FA05} @Defop LDVARINT32  // (s -- x s')
x{FA06} @Defop STVARUINT32 // (b x -- b')
x{FA07} @Defop STVARINT32  // (b x -- b')

x{FA40} @Defop LDMSGADDR
x{FA41} @Defop LDMSGADDRQ
x{FA42} @Defop PARSEMSGADDR
x{FA43} @Defop PARSEMSGADDRQ
x{FA44} @Defop REWRITESTDADDR
x{FA45} @Defop REWRITESTDADDRQ
x{FA46} @Defop REWRITEVARADDR
x{FA47} @Defop REWRITEVARADDRQ
x{FA48} @Defop LDSTDADDR
x{FA49} @Defop LDSTDADDRQ
x{FA50} @Defop LDOPTSTDADDR
x{FA51} @Defop LDOPTSTDADDRQ
x{FA52} @Defop STSTDADDR
x{FA53} @Defop STSTDADDRQ
x{FA54} @Defop STOPTSTDADDR
x{FA55} @Defop STOPTSTDADDRQ

x{FB00} @Defop SENDRAWMSG
x{FB02} @Defop RAWRESERVE
x{FB03} @Defop RAWRESERVEX
x{FB04} @Defop SETCODE
x{FB06} @Defop SETLIBCODE
x{FB07} @Defop CHANGELIB
x{FB08} @Defop SENDMSG

//
// debug primitives

{ dup 0 239 @-range abort"debug selector out of range"
  <b x{FE} s, swap 8 u, @addopb
} : DEBUG
{ dup $len 1- <b x{FEF} s, swap 4 u, swap $, @addopb
} : DEBUGSTR
{ over $len <b x{FEF} s, swap 4 u, swap 8 u, swap $, @addopb
} : DEBUGSTRI

x{FE00} @Defop DUMPSTK
{ 1 15 @rangechk <b x{FE0} s, swap 4 u, @addopb
} : DUMPSTKTOP
x{FE10} @Defop HEXDUMP
x{FE11} @Defop HEXPRINT
x{FE12} @Defop BINDUMP
x{FE13} @Defop BINPRINT
x{FE14} @Defop STRDUMP
x{FE15} @Defop STRPRINT
x{FE1E} @Defop DEBUGOFF
x{FE1F} @Defop DEBUGON
x{FE2} @Defop(s) DUMP
x{FE3} @Defop(s) PRINT
' DEBUGSTR : DUMPTOSFMT
{ 0 DEBUGSTRI } : LOGSTR
{ 1 DEBUGSTRI } : PRINTSTR
x{FEF000} @Defop LOGFLUSH

//
// codepage primitives
x{FF00} @Defop SETCP0
x{FFF0} @Defop SETCPX
{ dup -14 239 @-range abort"codepage out of range"
  255 and <b x{FF} s, swap 8 u, @addopb
} : SETCP

' @addop : CUSTOMOP

//
// provisions for defining programs consisting of several mutually-recursive procedures
//
variable @proccnt
variable @proclist
variable @procdict
variable @procinfo
variable @gvarcnt
variable @parent-state
variable asm-mode  1 asm-mode !
19 constant @procdictkeylen
32 constant @zcount
{ pair @proclist @ cons @proclist ! } : @proclistadd
{ @procinfo @ @procdictkeylen idict@ { 16 i@ } { 0 } cond } : @procinfo@
{ <b rot 16 i, swap @procinfo @ @procdictkeylen b>idict!
  not abort"cannot add key to procedure info dictionary"
  @procinfo !
} : @procinfo!
// ( x v1 v2 -- )
{ not 2 pick @procinfo@ and xor swap @procinfo! } : @procinfo~!
// ( s i f -- )
{ over @procdictkeylen fits not abort"procedure index out of range"
  over swap dup @procinfo~!  2dup @proclistadd
  1 'nop does swap 0 (create)
} : @declproc
{ 1 'nop does swap 0 (create) } : @declglobvar
{ @proccnt @ 1+ dup @proccnt ! 1 @declproc } : @newproc
{ @gvarcnt @ 1+ dup @gvarcnt ! @declglobvar } : @newglobvar
variable @oldcurrent  variable @oldctx
Fift-wordlist dup @oldcurrent ! @oldctx !
{ current@ @oldcurrent !  context@ @oldctx !  Asm definitions
  @proccnt @ @proclist @ @procdict @ @procinfo @ @gvarcnt @ @parent-state @ current@ @oldcurrent @ @oldctx @
  9 tuple @parent-state !
  hole current!
  0 =: main  @proclist null!  @proccnt 0!  @gvarcnt 0!
  { bl word @newproc } : NEWPROC
  { bl word dup (def?) ' drop ' @newproc cond } : DECLPROC
  { bl word dup find
    { nip execute <> abort"method redefined with different id" }
    { swap 17 @declproc }
  cond } : DECLMETHOD
  { bl word @newglobvar } : DECLGLOBVAR
  "main" 0 @proclistadd
  dictnew dup @procdict !
  @procinfo !  16 0 @procinfo!
} : PROGRAM{
{ over sbits < { s>c <b swap ref, b> <s } if } : @adj-long-proc
{ // i s l
  dup 0< {
    negate
    @was-split @ { drop 0 } if
  } if
  @adj-long-proc over @procdict @ @procdictkeylen
  idict!+ not abort"cannot define procedure, redefined?"
  @procdict !  2 2 @procinfo~!
} : @def-proc
{ @procinfo @ null? not } : @have-procinfo?
{ @have-procinfo? { 4 4 @procinfo~! } { drop } cond } : @proc-inlined
{ @have-procinfo? { 8 8 @procinfo~! } { drop } cond } : @proc-called
{ 1000 @def-proc } : PROC
{ 0 @def-proc } : PROCREF
{ -1000 @def-proc } : PROCINLINE
{ @procdict @ @procdictkeylen idict@ abort"procedure already defined"
} : @fail-ifdef
{ u@?+ { swap abort"first bits are not zeroes" } if } : @cut-zeroes
{ over @fail-ifdef
  2 { rot @normal? rot b> <s @zcount @cut-zeroes swap @def-proc drop } does
  null swap @doafter<{ 0 @zcount u,
} : @PROC:<{
{ 1000 @PROC:<{ } : PROC:<{
{ 0 @PROC:<{ } : PROCREF:<{
{ -1000 @PROC:<{ } : PROCINLINE:<{
{ dup @proc-called CALLDICT } dup : CALL : CALLDICT
{ dup @proc-called JMPDICT } dup : JMP : JMPDICT
{ dup @proc-called PREPAREDICT } dup : PREPARE : PREPAREDICT
{ dup @procdict @ @procdictkeylen idict@
  { swap @proc-inlined INLINE } { CALLDICT } cond
} dup : INLINECALL : INLINECALLDICT
{ 0 @procdict @ @procdictkeylen idict@ not abort"\`main\` procedure not defined" drop
} : @chkmaindef
{ @procdict @ @procdictkeylen idict- drop @procdict ! } : @remove-proc
{ ."Procedure \`" over type ."\` index=" 2 pick . ." flags=0x" dup x. cr } : @showprocinfo
// ( proc_name proc_idx f -- )  f:+1=declared, +2=defined, +4=inlined, +8=called, +16=method
{ // @showprocinfo
  dup 0x1a and 2 = asm-mode @ 3 and and ?dup {
    2 and {
      over ."Warning: removing (inlined) procedure \`" type ."\` from call dictionary" cr
    } if
    2 pick @remove-proc
  } if  // remove unused procs
  dup 0xc and 0xc = asm-mode @ 4 and and {
    over ."Warning: inline procedure \`" type ."\` is not always inlined" cr
  } if
  dup 0x1e and 2 = asm-mode @ 8 and and {
    over ."Warning: procedure \`" type ."\` defined but not used" cr
  } if
  drop 2drop
} : @chkprocdef
{ @chkmaindef
  @proclist @ { dup null? not } {
    uncons swap unpair over find not
    { drop +": undefined procedure name in list" abort } if
    drop tuck @procdict @ @procdictkeylen idict@ not
    { +": procedure declared but left undefined" abort } if
    drop swap 2dup @procinfo@ @chkprocdef (forget)
  } while
  drop @proclist null! @procinfo null! @proccnt 0!
  @procdict dup @ swap null!
  @parent-state @ dup null? { drop } {
    9 untuple
    @oldctx ! @oldcurrent ! current! @parent-state ! @gvarcnt ! @procinfo ! @procdict ! @proclist ! @proccnt !
  } cond
  @oldctx @ context!  @oldcurrent @ current!
} : }END
forget @proclist  forget @proccnt
{ }END <{ SETCP0 swap @procdictkeylen DICTPUSHCONST DICTIGETJMPZ 11 THROWARG }> } : }END>
{ }END> b> } : }END>c
{ }END>c <s } : }END>s

// This is the way how FunC assigns method_id for reserved functions.
// Note, that Tolk entrypoints have other names (\`onInternalMessage\`, etc.),
// but method_id is assigned not by Fift, but by Tolk code generation.
0 constant recv_internal
-1 constant recv_external
-2 constant run_ticktock
-3 constant split_prepare
-4 constant split_install

{ asm-mode 0 3 ~! } : asm-no-remove-unused
{ asm-mode 1 1 ~! } : asm-remove-unused   // enabled by default
{ asm-mode 3 3 ~! } : asm-warn-remove-unused
{ asm-mode 4 4 ~! } : asm-warn-inline-mix
{ asm-mode 0 4 ~! } : asm-no-warn-inline-mix  // disabled by default
{ asm-mode 8 8 ~! } : asm-warn-unused
{ asm-mode 0 8 ~! } : asm-no-warn-unused  // disabled by default

// ( c -- )   add vm library for later use with runvmcode
{ <b over ref, b> <s swap hash vmlibs @ 256 udict! not abort"cannot add library" vmlibs ! } : add-lib
// ( x -- c ) make library reference cell
{ <b 2 8 u, swap 256 u, b>spec } : hash>libref
// ( c -- c' )
{ hash hash>libref } : >libref

{ dup "." $pos dup -1 =
     { drop 0 }
     { $| 1 $| nip swap (number) 1- abort"invalid version"
       dup dup 0 < swap 999 > or abort"invalid version"
     }
     cond
} : parse-version-level

{
  0 swap
  "." $+
  { swap 1000 * swap parse-version-level rot + swap } 3 times
  "" $= not abort"invalid version"
} : parse-asm-fif-version

{
  dup =: required-version parse-asm-fif-version
  asm-fif-version parse-asm-fif-version
  = 1+ {
      "Required Asm.fif version: " @' required-version "; actual Asm.fif version: " asm-fif-version $+ $+ $+ abort
    } if
} : require-asm-fif-version

{
  dup =: required-version parse-asm-fif-version
  asm-fif-version parse-asm-fif-version
  swap
  >= 1+ {
      "Required Asm.fif version: " @' required-version "; actual Asm.fif version: " asm-fif-version $+ $+ $+ abort
    } if
} : require-asm-fif-version>=


Fift definitions Asm
' <{ : <{
' PROGRAM{ : PROGRAM{
' asm-fif-version : asm-fif-version
' require-asm-fif-version : require-asm-fif-version
' require-asm-fif-version>= : require-asm-fif-version>=
Fift
`,

'@fiftlib/Fift.fif':`{ 0 word drop 0 'nop } :: //
{ char " word 1 { swap { abort } if drop } } ::_ abort"
{ { bl word dup "" $= abort"comment extends after end of file" "*/" $= } until 0 'nop } :: /*
// { bl word 1 2 ' (create) } "::" 1 (create)
// { bl word 0 2 ' (create) } :: :
// { bl word 2 2 ' (create) } :: :_
// { bl word 3 2 ' (create) } :: ::_
// { bl word 0 (create) } : create
// { bl word (forget) } : forget
{ bl word 1 ' (forget) } :: [forget]
{ char " word 1 ' type } ::_ ."
{ char } word x>B 1 'nop } ::_ B{
{ swap ({) over 2+ -roll swap (compile) (}) } : does
{ 1 'nop does create } : constant
{ 2 'nop does create } : 2constant
{ hole constant } : variable
10 constant ten
{ bl word 1 { find 0= abort"word not found" } } :: (')
{ bl word find not abort"-?" 0 swap } :: [compile]
{ bl word 1 {
  dup find { " -?" $+ abort } ifnot nip execute
} } :: @'
{ bl word 1 { swap 1 'nop does swap 0 (create) }
} :: =:
{ bl word 1 { -rot 2 'nop does swap 0 (create) }
} :: 2=:
{ <b swap s, b> } : s>c
{ s>c hashB } : shash
// to be more efficiently re-implemented in C++ in the future
{ dup 0< ' negate if } : abs
{ 2dup > ' swap if } : minmax
{ minmax drop } : min
{ minmax nip } : max
"" constant <#
' $reverse : #>
{ swap 10 /mod char 0 + rot swap hold } : #
{ { # over 0<= } until } : #s
{ 0< { char - hold } if } : sign
// { dup abs <# #s rot sign #> nip } : (.)
// { (.) type } : ._
// { ._ space } : .
{ dup 10 < { 48 } { 55 } cond + } : Digit
{ dup 10 < { 48 } { 87 } cond + } : digit
// x s b -- x' s'
{ rot swap /mod Digit rot swap hold } : B#
{ rot swap /mod digit rot swap hold } : b#
{ 16 B# } : X#
{ 16 b# } : x#
// x s b -- 0 s'
{ -rot { 2 pick B# over 0<= } until rot drop } : B#s
{ -rot { 2 pick b# over 0<= } until rot drop } : b#s
{ 16 B#s } : X#s
{ 16 b#s } : x#s
variable base
{ 10 base ! } : decimal
{ 16 base ! } : hex
{ 8 base ! } : octal
{ 2 base ! } : binary
{ base @ B# } : Base#
{ base @ b# } : base#
{ base @ B#s } : Base#s
{ base @ b#s } : base#s
// x w -- s
{ over abs <# rot 1- ' X# swap times X#s rot sign #> nip } : (0X.)
{ over abs <# rot 1- ' x# swap times x#s rot sign #> nip } : (0x.)
{ (0X.) type } : 0X._
{ 0X._ space } : 0X.
{ (0x.) type } : 0x._
{ 0x._ space } : 0x.
{ bl (-trailing) } : -trailing
{ char 0 (-trailing) } : -trailing0
{ char " word 1 ' $+ } ::_ +"
{ find 0<> dup ' nip if } : (def?)
{ bl word 1 ' (def?) } :: def?
{ bl word 1 { (def?) not } } :: undef?
{ def? ' skip-to-eof if } : skip-ifdef
{ bl word dup (def?) { drop skip-to-eof } { 'nop swap 0 (create) } cond } : library
{ bl word dup (def?) { 2drop skip-to-eof } { swap 1 'nop does swap 0 (create) } cond } : library-version
{ hole dup 1 'nop does swap 1 { context! } does bl word tuck 0 (create) +"-wordlist" 0 (create) } : namespace
{ context@ current! } : definitions
{ char ) word "$" swap $+ 1 { find 0= abort"undefined parameter" execute } } ::_ $(
// b s -- ?
{ sbitrefs rot brembitrefs rot >= -rot <= and } : s-fits?
// b s x -- ?
{ swap sbitrefs -rot + rot brembitrefs -rot <= -rot <= and } : s-fits-with?
{ 0 swap ! } : 0!
{ tuck @ + swap ! } : +!
{ tuck @ swap - swap ! } : -!
{ 1 swap +! } : 1+!
{ -1 swap +! } : 1-!
{ null swap ! } : null!
{ not 2 pick @ and xor swap ! } : ~!
0 tuple constant nil
{ 1 tuple } : single
{ 2 tuple } : pair
{ 3 tuple } : triple
{ 1 untuple } : unsingle
{ 2 untuple } : unpair
{ 3 untuple } : untriple
{ over tuple? { swap count = } { 2drop false } cond } : tuple-len?
{ 0 tuple-len? } : nil?
{ 1 tuple-len? } : single?
{ 2 tuple-len? } : pair?
{ 3 tuple-len? } : triple?
{ 0 [] } : first
{ 1 [] } : second
{ 2 [] } : third
' pair : cons
' unpair : uncons
{ 0 [] } : car
{ 1 [] } : cdr
{ cdr car } : cadr
{ cdr cdr } : cddr
{ cdr cdr car } : caddr
{ null ' cons rot times } : list
{ -rot pair swap ! } : 2!
{ @ unpair } : 2@
{ true (atom) drop } : atom
{ bl word atom 1 'nop } ::_ \`
{ hole dup 1 { @ execute } does create } : recursive
{ 0 { 1+ dup 1 ' $() does over (.) "$" swap $+ 0 (create) } rot times drop } : :$1..n
{ 10 hold } : +cr
{ 9 hold } : +tab
{ "" swap { 0 word 2dup $cmp } { rot swap $+ +cr swap } while 2drop } : scan-until-word
{ 0 word -trailing scan-until-word 1 'nop } ::_ $<<
{ 0x40 runvmx } : runvmcode
{ 0x48 runvmx } : gasrunvmcode
{ 0xc8 runvmx } : gas2runvmcode
{ 0x43 runvmx } : runvmdict
{ 0x4b runvmx } : gasrunvmdict
{ 0xcb runvmx } : gas2runvmdict
{ 0x45 runvmx } : runvm
{ 0x4d runvmx } : gasrunvm
{ 0xcd runvmx } : gas2runvm
{ 0x55 runvmx } : runvmctx
{ 0x5d runvmx } : gasrunvmctx
{ 0xdd runvmx } : gas2runvmctx
{ 0x75 runvmx } : runvmctxact
{ 0x7d runvmx } : gasrunvmctxact
{ 0xfd runvmx } : gas2runvmctxact
{ 0x35 runvmx } : runvmctxactq
{ 0x3d runvmx } : gasrunvmctxactq
`,

};
